// igniteui-webcomponents-gauges/esm2015/lib/igc-radial-gauge-component.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebComponentRenderer } from "igniteui-webcomponents-core";
import { TypeRegistrar } from "igniteui-webcomponents-core";
import { XamRadialGauge } from './XamRadialGauge';
import { CollectionAdapter, brushToString, stringToBrush, toPoint, fromPoint, fromBrushCollection, toBrushCollection, ensureBool, ensureEnum, initializePropertiesFromCss, NamePatcher, toSpinal, fromSpinal, enumToString, brushCollectionToString, getAllPropertyNames, arrayFindByName } from "igniteui-webcomponents-core";
import { delegateCombine, delegateRemove } from "igniteui-webcomponents-core";
import { NotifyCollectionChangedAction } from "igniteui-webcomponents-core";
import { SweepDirection_$type } from "igniteui-webcomponents-core";
import { RadialGaugeNeedleShape_$type } from './RadialGaugeNeedleShape';
import { RadialGaugePivotShape_$type } from './RadialGaugePivotShape';
import { RadialGaugeScaleOversweepShape_$type } from './RadialGaugeScaleOversweepShape';
import { RadialGaugeBackingShape_$type } from './RadialGaugeBackingShape';
import { RadialGaugeDuplicateLabelOmissionStrategy_$type } from './RadialGaugeDuplicateLabelOmissionStrategy';
import { IgcFormatRadialGaugeLabelEventArgs } from './igc-format-radial-gauge-label-event-args';
import { IgcAlignRadialGaugeLabelEventArgs } from './igc-align-radial-gauge-label-event-args';
import { RadialGaugeStylingDefaults } from './RadialGaugeStylingDefaults';
import { IgcRadialGaugeRangeCollection } from './igc-radial-gauge-range-collection';
import { IgcRadialGaugeRangeComponent } from './igc-radial-gauge-range-component';
import { RegisterElementHelper } from "igniteui-webcomponents-core";
import { IgcDoubleValueChangedEventArgs } from "igniteui-webcomponents-core";
import { IgcHTMLElement } from "igniteui-webcomponents-core";
export let IgcRadialGaugeComponent = /*@__PURE__*/ (() => {
    class IgcRadialGaugeComponent extends IgcHTMLElement {
        constructor() {
            super();
            this.contentRanges = [];
            /**
            * The ranges actually present in the chart. Do not directly modify this array.
            * This array's contents can be modified by causing Angular to reproject the child content.
            * Or adding and removing ranges from the manual ranges collection on the ranges property.
            */
            this.actualRanges = [];
            this._ranges = null;
            this._rangesAdapter = null;
            this._disconnected = false;
            this._settingAttributes = false;
            this._attached = false;
            this._queuedSetAttributes = [];
            this._updatingFromAttribute = false;
            this.__p = null;
            this._hasUserValues = new Set();
            this._stylingContainer = null;
            this._stylingParent = null;
            this._inStyling = false;
            this._formatLabel = null;
            this._formatLabel_wrapped = null;
            this._alignLabel = null;
            this._alignLabel_wrapped = null;
            this._valueChanged = null;
            this._valueChanged_wrapped = null;
            this._actualMinimumValueChange = null;
            this._actualMinimumValueChange_wrapped = null;
            this._actualMaximumValueChange = null;
            this._actualMaximumValueChange_wrapped = null;
            if (this._styling) {
                NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
            }
            this._renderer = new WebComponentRenderer(this, document, true, RadialGaugeStylingDefaults);
            this._implementation = this.createImplementation();
            this._container = this._renderer.createElement("div");
            this._renderer.updateRoot(this._container);
            //this._renderer.rootWrapper.append(this._container);
            this._container.setStyleProperty("display", "block");
            this._container.setStyleProperty("width", "100%");
            this._container.setStyleProperty("height", "100%");
            var root;
            root = this._container;
            if (this._container.getNativeElement() != null) {
                root = this._container.getNativeElement();
            }
            this._wrapper = this._renderer;
            var gauge = this.i;
            this._gauge = gauge;
            this._rangesAdapter = new CollectionAdapter(this.contentRanges, this.i.ranges, this.actualRanges, (c) => c.i, (i) => {
                i._provideRenderer(this._wrapper);
                if (this.parentElement) {
                    i._styling(this, this, this);
                }
            }, (i) => { i._provideRenderer(null); });
            gauge.provideContainer(this._renderer);
            let mut = new MutationObserver((list) => {
                for (var mutation of list) {
                    if (mutation.type == 'childList') {
                        this.updateContentRanges();
                    }
                }
            });
            mut.observe(this, {
                childList: true
            });
            this._renderer.addSizeWatcher(() => {
                this._gauge.containerResized();
            });
        }
        set height(value) {
            this._height = value;
            this.style.height = value;
            this.containerResized();
        }
        get height() {
            return this._height;
        }
        set width(value) {
            this._width = value;
            this.style.width = value;
            this.containerResized();
        }
        get width() {
            return this._width;
        }
        /**
         * A collection or manually added axes for the chart.
        */
        get ranges() {
            if (this._ranges === null) {
                let coll = new IgcRadialGaugeRangeCollection();
                let inner = coll._innerColl;
                inner.addListener((sender, e) => {
                    switch (e.action) {
                        case NotifyCollectionChangedAction.Add:
                            this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Remove:
                            this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            break;
                        case NotifyCollectionChangedAction.Replace:
                            this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Reset:
                            this._rangesAdapter.clearManualItems();
                            break;
                    }
                });
                this._ranges = coll;
            }
            return this._ranges;
        }
        destroy() {
            this._gauge.destroy();
            this._wrapper.destroy();
        }
        updateContentRanges() {
            this.contentRanges.length = 0;
            for (var i = 0; i < this.children.length; i++) {
                if (this.children[i] instanceof IgcRadialGaugeRangeComponent) {
                    this.contentRanges.push(this.children[i]);
                }
            }
            if (this._rangesAdapter != null) {
                this._rangesAdapter.notifyContentChanged();
            }
        }
        createImplementation() {
            return new XamRadialGauge();
        }
        disconnectedCallback() {
            this._disconnected = true;
            if (this.i) {
                this.i.onDetachedFromUI();
            }
        }
        connectedCallback() {
            if (this._disconnected) {
                this._disconnected = false;
                if (this.i) {
                    this.i.onAttachedToUI();
                }
                return;
            }
            this.classList.add("ig-radial-gauge");
            this.classList.add("igc-radial-gauge");
            this.appendChild(this._renderer.rootWrapper.getNativeElement());
            this._attached = true;
            this.style.display = "block";
            this.style.height = this._height;
            this.style.width = this._width;
            this._flushQueuedAttributes();
            // supports themes or custom properties set in CSS 
            this._styling(this, this);
            for (let i = 0; i < this.actualRanges.length; i++) {
                let s = this.actualRanges[i];
                s._styling(this, this, this);
            }
            this.updateContentRanges();
            this.afterContentInit();
        }
        afterContentInit() {
            this.i.containerResized();
        }
        /**
         * @hidden
         */
        get i() {
            return this._implementation;
        } /**
                                     * @hidden
                                     */
        static _createFromInternal(internal) {
            if (!internal) {
                return null;
            }
            if (!internal.$type) {
                return null;
            }
            let name = internal.$type.name;
            let externalName = "Igc" + name + "Component";
            if (!TypeRegistrar.isRegistered(externalName)) {
                return null;
            }
            return TypeRegistrar.create(externalName);
        }
        _enqueueSetAttribute(attrName, attrValue) {
            this._queuedSetAttributes.push({ attrName: attrName, attrValue: attrValue });
        }
        _flushQueuedAttributes() {
            this._settingAttributes = true;
            for (let i = 0; i < this._queuedSetAttributes.length; i++) {
                this.setAttribute(this._queuedSetAttributes[i].attrName, this._queuedSetAttributes[i].attrValue);
            }
            this._settingAttributes = false;
            this._queuedSetAttributes.length = 0;
        }
        _a(attrName, attrValue) {
            if (this._updatingFromAttribute) {
                return;
            }
            if (attrValue) {
                attrValue = attrValue.toString();
            }
            this._settingAttributes = true;
            attrName = toSpinal(attrName);
            if (this._attached) {
                this.setAttribute(attrName, attrValue);
            }
            else {
                this._enqueueSetAttribute(attrName, attrValue);
            }
            this._settingAttributes = false;
        }
        static get observedAttributes() {
            if (IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent == null) {
                let names = getAllPropertyNames(IgcRadialGaugeComponent);
                for (let i = 0; i < names.length; i++) {
                    names[i] = toSpinal(names[i]);
                }
                IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent = names;
            }
            return IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent;
        }
        attributeChangedCallback(name, oldValue, newValue) {
            if (this._settingAttributes) {
                return;
            }
            let setName = fromSpinal(name);
            this._updatingFromAttribute = true;
            this[setName] = newValue;
            this._updatingFromAttribute = false;
        }
        static register() {
            if (!IgcRadialGaugeComponent._isElementRegistered) {
                IgcRadialGaugeComponent._isElementRegistered = true;
                RegisterElementHelper.registerElement(IgcRadialGaugeComponent.htmlTagName, IgcRadialGaugeComponent);
            }
        }
        /**
     * Gets or sets a collection of brushes to be used as the palette for gauge ranges.
    */
        get rangeBrushes() {
            return fromBrushCollection(this.i.ah);
        }
        set rangeBrushes(v) {
            this.i.ah = toBrushCollection(v);
            this._a("rangeBrushes", brushCollectionToString(this.i.ah));
        }
        /**
         * Gets or sets a collection of brushes to be used as the palette for gauge outlines.
        */
        get rangeOutlines() {
            return fromBrushCollection(this.i.ai);
        }
        set rangeOutlines(v) {
            this.i.ai = toBrushCollection(v);
            this._a("rangeOutlines", brushCollectionToString(this.i.ai));
        }
        /**
         * Gets or sets the minimum value of the scale.
        */
        get minimumValue() {
            return this.i.cc;
        }
        set minimumValue(v) {
            this.i.cc = +v;
            this._a("minimumValue", this.i.cc);
        }
        /**
         * Gets the resolved minimum value of the scale.
        */
        get actualMinimumValue() {
            return this.i.bs;
        }
        set actualMinimumValue(v) {
            this.i.bs = +v;
            this._a("actualMinimumValue", this.i.bs);
        }
        /**
         * Gets or sets the maximum value of the scale.
        */
        get maximumValue() {
            return this.i.ca;
        }
        set maximumValue(v) {
            this.i.ca = +v;
            this._a("maximumValue", this.i.ca);
        }
        /**
         * Gets the resolved maximum value of the scale.
        */
        get actualMaximumValue() {
            return this.i.br;
        }
        set actualMaximumValue(v) {
            this.i.br = +v;
            this._a("actualMaximumValue", this.i.br);
        }
        /**
         * Gets or sets the interval to use for the scale.
        */
        get interval() {
            return this.i.b6;
        }
        set interval(v) {
            this.i.b6 = +v;
            this._a("interval", this.i.b6);
        }
        /**
         * Gets or sets the x position of the center of the gauge with the value ranging from 0 to 1.
        */
        get centerX() {
            return this.i.b3;
        }
        set centerX(v) {
            this.i.b3 = +v;
            this._a("centerX", this.i.b3);
        }
        /**
         * Gets or sets the y position of the center of the gauge with the value ranging from 0 to 1.
        */
        get centerY() {
            return this.i.b4;
        }
        set centerY(v) {
            this.i.b4 = +v;
            this._a("centerY", this.i.b4);
        }
        /**
         * Gets or sets the value at which to point the needle of the gauge.
        */
        get value() {
            return this.i.dm;
        }
        set value(v) {
            this.i.dm = +v;
            this._a("value", this.i.dm);
        }
        /**
         * Gets or sets the start angle for the scale in degrees.
        */
        get scaleStartAngle() {
            return this.i.da;
        }
        set scaleStartAngle(v) {
            this.i.da = +v;
            this._a("scaleStartAngle", this.i.da);
        }
        /**
         * Gets or sets the end angle for the scale in degrees.
        */
        get scaleEndAngle() {
            return this.i.c5;
        }
        set scaleEndAngle(v) {
            this.i.c5 = +v;
            this._a("scaleEndAngle", this.i.c5);
        }
        /**
         * Gets or sets the direction in which the scale sweeps around the center from the start angle to end angle.
        */
        get scaleSweepDirection() {
            return this.i.ii;
        }
        set scaleSweepDirection(v) {
            this.i.ii = ensureEnum(SweepDirection_$type, v);
            this._a("scaleSweepDirection", enumToString(SweepDirection_$type, this.i.ii));
        }
        /**
         * Gets or sets the number of milliseconds over which changes to the gauge should be animated.
        */
        get transitionDuration() {
            return this.i.dn;
        }
        set transitionDuration(v) {
            this.i.dn = +v;
            this._a("transitionDuration", this.i.dn);
        }
        /**
         * Gets or sets the easing function used to morph the current series.
        */
        get transitionEasingFunction() {
            return this.i.a0;
        }
        set transitionEasingFunction(v) {
            this.i.a0 = v;
        }
        /**
         * Gets or sets the brush to use when rendering the fill of the needle.
        */
        get needleBrush() {
            return brushToString(this.i.ib);
        }
        set needleBrush(v) {
            this.i.ib = stringToBrush(v);
            this._a("needleBrush", brushToString(this.i.ib));
        }
        /**
         * Gets or sets the brush to use when rendering the outline of the needle.
        */
        get needleOutline() {
            return brushToString(this.i.ic);
        }
        set needleOutline(v) {
            this.i.ic = stringToBrush(v);
            this._a("needleOutline", brushToString(this.i.ic));
        }
        /**
         * Gets or sets the extent (from -1 to 1) at which to start rendering the needle, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleStartExtent() {
            return this.i.cv;
        }
        set needleStartExtent(v) {
            this.i.cv = +v;
            this._a("needleStartExtent", this.i.cv);
        }
        /**
         * Gets or sets the extent (from -1 to 1) at which to end rendering the needle, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleEndExtent() {
            return this.i.cn;
        }
        set needleEndExtent(v) {
            this.i.cn = +v;
            this._a("needleEndExtent", this.i.cn);
        }
        /**
         * Gets or sets the shape to use when rendering the needle from a number of options.
        */
        get needleShape() {
            return this.i.ar;
        }
        set needleShape(v) {
            this.i.ar = ensureEnum(RadialGaugeNeedleShape_$type, v);
            this._a("needleShape", enumToString(RadialGaugeNeedleShape_$type, this.i.ar));
        }
        /**
         * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleStartWidthRatio() {
            return this.i.cw;
        }
        set needleStartWidthRatio(v) {
            this.i.cw = +v;
            this._a("needleStartWidthRatio", this.i.cw);
        }
        /**
         * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleEndWidthRatio() {
            return this.i.co;
        }
        set needleEndWidthRatio(v) {
            this.i.co = +v;
            this._a("needleEndWidthRatio", this.i.co);
        }
        /**
         * Gets or sets the width of the needle at its feature which is closest to the base (e.g. a bulb) with a value from 0 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleBaseFeatureWidthRatio() {
            return this.i.cm;
        }
        set needleBaseFeatureWidthRatio(v) {
            this.i.cm = +v;
            this._a("needleBaseFeatureWidthRatio", this.i.cm);
        }
        /**
         * Gets or sets the extent of the feature which is closest to the base (e.g. a bulb) with a value from -1 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needleBaseFeatureExtent() {
            return this.i.cl;
        }
        set needleBaseFeatureExtent(v) {
            this.i.cl = +v;
            this._a("needleBaseFeatureExtent", this.i.cl);
        }
        /**
         * Gets or sets the width of the needle at its feature which is closest to the point (e.g. the tapering point of a needle) with a value from 0 to 1. Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needlePointFeatureWidthRatio() {
            return this.i.cu;
        }
        set needlePointFeatureWidthRatio(v) {
            this.i.cu = +v;
            this._a("needlePointFeatureWidthRatio", this.i.cu);
        }
        /**
         * Gets or sets the extent of the feature which is closest to the point (e.g. the tapering point of a needle) with a value from -1 to 1. Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needlePointFeatureExtent() {
            return this.i.ct;
        }
        set needlePointFeatureExtent(v) {
            this.i.ct = +v;
            this._a("needlePointFeatureExtent", this.i.ct);
        }
        /**
         * Gets or sets the width of the cap of the needle with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get needlePivotWidthRatio() {
            return this.i.cs;
        }
        set needlePivotWidthRatio(v) {
            this.i.cs = +v;
            this._a("needlePivotWidthRatio", this.i.cs);
        }
        /**
         * Gets or sets the width of the inner cutout section of the needle cap with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle that has a cutout section.
        */
        get needlePivotInnerWidthRatio() {
            return this.i.cp;
        }
        set needlePivotInnerWidthRatio(v) {
            this.i.cp = +v;
            this._a("needlePivotInnerWidthRatio", this.i.cp);
        }
        /**
         * Gets or sets the shape to use for the needle cap.
        */
        get needlePivotShape() {
            return this.i.as;
        }
        set needlePivotShape(v) {
            this.i.as = ensureEnum(RadialGaugePivotShape_$type, v);
            this._a("needlePivotShape", enumToString(RadialGaugePivotShape_$type, this.i.as));
        }
        /**
         * Gets or sets the position at which to start rendering the scale, measured from the center of the gauge as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get scaleStartExtent() {
            return this.i.db;
        }
        set scaleStartExtent(v) {
            this.i.db = +v;
            this._a("scaleStartExtent", this.i.db);
        }
        /**
         * Gets or sets the brush to use for filling the needle cap. Note: this only applies to certain cap shapes.
        */
        get needlePivotBrush() {
            return brushToString(this.i.id);
        }
        set needlePivotBrush(v) {
            this.i.id = stringToBrush(v);
            this._a("needlePivotBrush", brushToString(this.i.id));
        }
        /**
         * Gets or sets the brush to use for the outlines of the needle cap.
        */
        get needlePivotOutline() {
            return brushToString(this.i.ie);
        }
        set needlePivotOutline(v) {
            this.i.ie = stringToBrush(v);
            this._a("needlePivotOutline", brushToString(this.i.ie));
        }
        /**
         * Gets or sets the stroke thickness of the needle outline.
        */
        get needleStrokeThickness() {
            return this.i.cx;
        }
        set needleStrokeThickness(v) {
            this.i.cx = +v;
            this._a("needleStrokeThickness", this.i.cx);
        }
        /**
         * Gets or sets the stroke thickness to use for the outline of the needle cap.
        */
        get needlePivotStrokeThickness() {
            return this.i.cq;
        }
        set needlePivotStrokeThickness(v) {
            this.i.cq = +v;
            this._a("needlePivotStrokeThickness", this.i.cq);
        }
        /**
         * Gets or sets the position at which to stop rendering the scale as a value from 0 to 1 measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get scaleEndExtent() {
            return this.i.c6;
        }
        set scaleEndExtent(v) {
            this.i.c6 = +v;
            this._a("scaleEndExtent", this.i.c6);
        }
        /**
         * Gets or sets the position at which to put the labels as a value from 0 to 1, measured form the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get labelExtent() {
            return this.i.b7;
        }
        set labelExtent(v) {
            this.i.b7 = +v;
            this._a("labelExtent", this.i.b7);
        }
        /**
         * Gets or sets the interval to use for rendering labels. This defaults to be the same interval as the tickmarks on the scale.
        */
        get labelInterval() {
            return this.i.b9;
        }
        set labelInterval(v) {
            this.i.b9 = +v;
            this._a("labelInterval", this.i.b9);
        }
        /**
         * Gets or sets the position at which to start rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get tickStartExtent() {
            return this.i.dg;
        }
        set tickStartExtent(v) {
            this.i.dg = +v;
            this._a("tickStartExtent", this.i.dg);
        }
        /**
         * Gets or sets the position at which to stop rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get tickEndExtent() {
            return this.i.de;
        }
        set tickEndExtent(v) {
            this.i.de = +v;
            this._a("tickEndExtent", this.i.de);
        }
        /**
         * Gets or sets the stroke thickness to use when rendering ticks.
        */
        get tickStrokeThickness() {
            return this.i.di;
        }
        set tickStrokeThickness(v) {
            this.i.di = +v;
            this._a("tickStrokeThickness", this.i.di);
        }
        /**
         * Gets or sets the brush to use for the major tickmarks.
        */
        get tickBrush() {
            return brushToString(this.i.ih);
        }
        set tickBrush(v) {
            this.i.ih = stringToBrush(v);
            this._a("tickBrush", brushToString(this.i.ih));
        }
        /**
         * Gets or sets the brush to use for the label font.
        */
        get fontBrush() {
            return brushToString(this.i.h8);
        }
        set fontBrush(v) {
            this.i.h8 = stringToBrush(v);
            this._a("fontBrush", brushToString(this.i.h8));
        }
        /**
         * Gets or sets the position at which to start rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get minorTickStartExtent() {
            return this.i.ch;
        }
        set minorTickStartExtent(v) {
            this.i.ch = +v;
            this._a("minorTickStartExtent", this.i.ch);
        }
        /**
         * Gets or sets the position at which to stop rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get minorTickEndExtent() {
            return this.i.cf;
        }
        set minorTickEndExtent(v) {
            this.i.cf = +v;
            this._a("minorTickEndExtent", this.i.cf);
        }
        /**
         * Gets or sets the stroke thickness to use when rendering minor ticks.
        */
        get minorTickStrokeThickness() {
            return this.i.cj;
        }
        set minorTickStrokeThickness(v) {
            this.i.cj = +v;
            this._a("minorTickStrokeThickness", this.i.cj);
        }
        /**
         * Gets or sets the brush to use for the minor tickmarks.
        */
        get minorTickBrush() {
            return brushToString(this.i.ia);
        }
        set minorTickBrush(v) {
            this.i.ia = stringToBrush(v);
            this._a("minorTickBrush", brushToString(this.i.ia));
        }
        /**
         * Gets or sets the number of minor tickmarks to place between major tickmarks.
        */
        get minorTickCount() {
            return this.i.ce;
        }
        set minorTickCount(v) {
            this.i.ce = +v;
            this._a("minorTickCount", this.i.ce);
        }
        /**
         * Gets or sets the brush to use to fill the background of the scale.
        */
        get scaleBrush() {
            return brushToString(this.i.ig);
        }
        set scaleBrush(v) {
            this.i.ig = stringToBrush(v);
            this._a("scaleBrush", brushToString(this.i.ig));
        }
        /**
         * Gets or sets the brush to use to fill the backing of the gauge.
        */
        get backingBrush() {
            return brushToString(this.i.h6);
        }
        set backingBrush(v) {
            this.i.h6 = stringToBrush(v);
            this._a("backingBrush", brushToString(this.i.h6));
        }
        /**
         * Gets or sets the brush to use for the outline of the backing.
        */
        get backingOutline() {
            return brushToString(this.i.h7);
        }
        set backingOutline(v) {
            this.i.h7 = stringToBrush(v);
            this._a("backingOutline", brushToString(this.i.h7));
        }
        /**
         * Gets or sets the stroke thickness of the backing outline.
        */
        get backingStrokeThickness() {
            return this.i.b1;
        }
        set backingStrokeThickness(v) {
            this.i.b1 = +v;
            this._a("backingStrokeThickness", this.i.b1);
        }
        /**
         * Gets or sets the outer extent of the gauge backing.
        */
        get backingOuterExtent() {
            return this.i.by;
        }
        set backingOuterExtent(v) {
            this.i.by = +v;
            this._a("backingOuterExtent", this.i.by);
        }
        /**
         * Gets or sets the over sweep angle to apply to the backing if it is displaying fitted (in degrees). Must be greater or equal to 0.
        */
        get backingOversweep() {
            return this.i.b0;
        }
        set backingOversweep(v) {
            this.i.b0 = +v;
            this._a("backingOversweep", this.i.b0);
        }
        /**
         * Gets or sets the extra degrees of sweep to apply to the scale background. Must be greater or equal to 0.
        */
        get scaleOversweep() {
            return this.i.c8;
        }
        set scaleOversweep(v) {
            this.i.c8 = +v;
            this._a("scaleOversweep", this.i.c8);
        }
        /**
         * Gets or sets the oversweep shape to use for the excess fill area for the scale.
        */
        get scaleOversweepShape() {
            return this.i.au;
        }
        set scaleOversweepShape(v) {
            this.i.au = ensureEnum(RadialGaugeScaleOversweepShape_$type, v);
            this._a("scaleOversweepShape", enumToString(RadialGaugeScaleOversweepShape_$type, this.i.au));
        }
        /**
         * Gets or sets the corner rounding radius to use for the fitted scale backings.
        */
        get backingCornerRadius() {
            return this.i.bu;
        }
        set backingCornerRadius(v) {
            this.i.bu = +v;
            this._a("backingCornerRadius", this.i.bu);
        }
        /**
         * Gets or sets the inner extent of the gauge backing.
        */
        get backingInnerExtent() {
            return this.i.bw;
        }
        set backingInnerExtent(v) {
            this.i.bw = +v;
            this._a("backingInnerExtent", this.i.bw);
        }
        /**
         * Gets or sets the type of shape to use for the backing of the gauge.
        */
        get backingShape() {
            return this.i.am;
        }
        set backingShape(v) {
            this.i.am = ensureEnum(RadialGaugeBackingShape_$type, v);
            this._a("backingShape", enumToString(RadialGaugeBackingShape_$type, this.i.am));
        }
        /**
         * Gets or sets the multiplying factor to apply to the normal radius of the gauge.
         * The radius of the gauge is defined by the minimum of the width and height of the control divided by 2.0.
         * This introduces a multiplicative factor to that value.
        */
        get radiusMultiplier() {
            return this.i.c0;
        }
        set radiusMultiplier(v) {
            this.i.c0 = +v;
            this._a("radiusMultiplier", this.i.c0);
        }
        /**
         * Gets or sets the strategy to use for omitting labels if the first and last label have the same value.
        */
        get duplicateLabelOmissionStrategy() {
            return this.i.an;
        }
        set duplicateLabelOmissionStrategy(v) {
            this.i.an = ensureEnum(RadialGaugeDuplicateLabelOmissionStrategy_$type, v);
            this._a("duplicateLabelOmissionStrategy", enumToString(RadialGaugeDuplicateLabelOmissionStrategy_$type, this.i.an));
        }
        /**
         * Gets or sets whether needle dragging is enabled or not.
        */
        get isNeedleDraggingEnabled() {
            return this.i.ba;
        }
        set isNeedleDraggingEnabled(v) {
            this.i.ba = ensureBool(v);
            this._a("isNeedleDraggingEnabled", this.i.ba);
        }
        /**
         * Gets or sets whether the needle is constrained within the minimum and maximum value range during dragging.
        */
        get isNeedleDraggingConstrained() {
            return this.i.a9;
        }
        set isNeedleDraggingConstrained(v) {
            this.i.a9 = ensureBool(v);
            this._a("isNeedleDraggingConstrained", this.i.a9);
        }
        /**
         * Gets or sets the font.
        */
        get font() {
            return this.i.d2;
        }
        set font(v) {
            this.i.d2 = v;
        }
        /**
         * Gets or sets the label composite format used when creating label values.
        */
        get labelFormat() {
            return this.i.d9;
        }
        set labelFormat(v) {
            this.i.d9 = v;
        }
        /**
         * Gets or sets the format specifiers to use with the LabelFormat string.
        */
        get labelFormatSpecifiers() {
            return this.i.ag;
        }
        set labelFormatSpecifiers(v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                const re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ag = v;
        }
        /**
         * Gets a value indicating whether the radial gauge is currently animating.
        */
        get animating() {
            return this.i.a6;
        }
        /**
         * Gets the transition progress of the animation when the control is animating.
        */
        get transitionProgress() {
            return this.i.dk;
        }
        set transitionProgress(v) {
            this.i.dk = +v;
            this._a("transitionProgress", this.i.dk);
        }
        /**
         * Gets or sets the scaling value used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get pixelScalingRatio() {
            return this.i.cz;
        }
        set pixelScalingRatio(v) {
            this.i.cz = +v;
            this._a("pixelScalingRatio", this.i.cz);
        }
        /**
         * Gets the actual pixel scaling ratio used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get actualPixelScalingRatio() {
            return this.i.bt;
        }
        set actualPixelScalingRatio(v) {
            this.i.bt = +v;
            this._a("actualPixelScalingRatio", this.i.bt);
        }
        findByName(name) {
            if (this.findEphemera) {
                if (name && name.indexOf("@@e:") == 0) {
                    return this.findEphemera(name);
                }
            }
            if (this.ranges != null && this.ranges.findByName && this.ranges.findByName(name)) {
                return this.ranges.findByName(name);
            }
            if (this.labelFormatSpecifiers != null && arrayFindByName(this.labelFormatSpecifiers, name)) {
                return arrayFindByName(this.labelFormatSpecifiers, name);
            }
            return null;
        }
        get hasUserValues() {
            return this._hasUserValues;
        }
        __m(propertyName) {
            if (!this._inStyling) {
                this._hasUserValues.add(propertyName);
            }
        }
        _styling(container, component, parent) {
            if (this._inStyling) {
                return;
            }
            this._inStyling = true;
            this._stylingContainer = container;
            this._stylingParent = component;
            let genericPrefix = "";
            let typeName = this.i.$type.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            genericPrefix = toSpinal("RadialGaugeComponent");
            let additionalPrefixes = [];
            let prefix = toSpinal(typeName);
            additionalPrefixes.push(prefix + "-");
            let b = this.i.$type.baseType;
            while (b && b.name != "Object" &&
                b.name != "Base" &&
                b.name != "Control" &&
                b.Name != "DependencyObject" &&
                b.Name != "FrameworkElement") {
                typeName = b.name;
                if (typeName.indexOf("Xam") === 0) {
                    typeName = typeName.substring(3);
                }
                let basePrefix = toSpinal(typeName);
                additionalPrefixes.push(basePrefix + "-");
                b = b.baseType;
            }
            if (parent) {
                let parentTypeName = parent.i.$type.name;
                if (parentTypeName.indexOf("Xam") === 0) {
                    parentTypeName = parentTypeName.substring(3);
                }
                let parentPrefix = toSpinal(parentTypeName);
                additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
                additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
            }
            initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
            if (this._otherStyling) {
                this._otherStyling(container, component, parent);
            }
            this._inStyling = false;
        }
        /**
     * Scales a value on the gauge's main scale to an angle around the center point of the gauge, in radians.
    
    * @param value  * The value to scale.
    */
        scaleValue(value) {
            let iv = this.i.dd(value);
            return (iv);
        }
        /**
         * Unscales a value from an angle in radians to the represented value along the main scale of the gauge.
        
        * @param angle  * The angle in radians to unscale to a value.
        */
        unscaleValue(angle) {
            let iv = this.i.dl(angle);
            return (iv);
        }
        /**
         * Gets the value for the main scale of the gauge for a given point within the bounds of the gauge.
        
        * @param point  * The point for which to retrieve the associated value.
        */
        getValueForPoint(point) {
            let iv = this.i.b5(toPoint(point));
            return (iv);
        }
        getPointForValue(value, extent) {
            let iv = this.i.ik(value, extent);
            return fromPoint(iv);
        }
        styleUpdated() {
            this.i.gd();
        }
        needleContainsPoint(point, isFinger) {
            let iv = this.i.needleContainsPoint(toPoint(point), isFinger);
            return (iv);
        }
        /**
         * Called by the UI framework to provide a UI container for rendering this control.
        
        * @param container  * The UI container element.
        */
        provideContainer(container) {
            this.i.provideContainer(container);
        }
        containerResized() {
            this.i.containerResized();
        }
        /**
         * Use to force the radial gauge to finish any deferred work before printing or evaluating its visual.
         * This should only be called if the visual of the radial gauge needs to be synchronously saved or evaluated.
         * Calling this method too often will hinder the performance of the radial gauge.
        
        */
        flush() {
            this.i.fs();
        }
        /**
         * Returns visuals as a serialized string.
        
        */
        exportSerializedVisualData() {
            let iv = this.i.d1();
            return (iv);
        }
        /**
         * Event which is raised when a label of the gauge is formatted.
         * Function takes first argument null and second argument ui.
         * Use ui.owner to obtain reference to gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
         * Use ui.startAngle to obtain the starting angle of gauge scale.
         * Use ui.endAngle to obtain the ending angle of gauge scale.
         * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
         * Use ui.value to obtain the value on the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
        */
        get formatLabel() {
            return this._formatLabel;
        }
        set formatLabel(ev) {
            if (this._formatLabel_wrapped !== null) {
                this.i.formatLabel = delegateRemove(this.i.formatLabel, this._formatLabel_wrapped);
                this._formatLabel_wrapped = null;
                this._formatLabel = null;
            }
            this._formatLabel = ev;
            this._formatLabel_wrapped = (o, e) => {
                let outerArgs = new IgcFormatRadialGaugeLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (this.beforeFormatLabel) {
                    this.beforeFormatLabel(this, outerArgs);
                }
                if (this._formatLabel) {
                    this._formatLabel(this, outerArgs);
                }
            };
            this.i.formatLabel = delegateCombine(this.i.formatLabel, this._formatLabel_wrapped);
            ;
        }
        /**
         * Event which is raised when a label of the gauge is aligned along the scale.
         * Function takes first argument null and second argument ui.
         * Use ui.owner to obtain reference to gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
         * Use ui.startAngle to obtain the starting angle of gauge scale.
         * Use ui.endAngle to obtain the ending angle of gauge scale.
         * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
         * Use ui.value to obtain the value on the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
         * Use ui.width to obtain the width of the label.
         * Use ui.height to obtain the height of the label.
         * Use ui.offsetX to obtain the X offset of the label on the gauge scale.
         * Use ui.offsetY to obtain the Y offset of the label on the gauge scale.
        */
        get alignLabel() {
            return this._alignLabel;
        }
        set alignLabel(ev) {
            if (this._alignLabel_wrapped !== null) {
                this.i.alignLabel = delegateRemove(this.i.alignLabel, this._alignLabel_wrapped);
                this._alignLabel_wrapped = null;
                this._alignLabel = null;
            }
            this._alignLabel = ev;
            this._alignLabel_wrapped = (o, e) => {
                let outerArgs = new IgcAlignRadialGaugeLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (this.beforeAlignLabel) {
                    this.beforeAlignLabel(this, outerArgs);
                }
                if (this._alignLabel) {
                    this._alignLabel(this, outerArgs);
                }
            };
            this.i.alignLabel = delegateCombine(this.i.alignLabel, this._alignLabel_wrapped);
            ;
        }
        /**
         * Occurs when the Value property changes.
        */
        get valueChanged() {
            return this._valueChanged;
        }
        set valueChanged(ev) {
            if (this._valueChanged_wrapped !== null) {
                this.i.valueChanged = delegateRemove(this.i.valueChanged, this._valueChanged_wrapped);
                this._valueChanged_wrapped = null;
                this._valueChanged = null;
            }
            this._valueChanged = ev;
            this._valueChanged_wrapped = (o, e) => {
                let outerArgs = new IgcDoubleValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (this.beforeValueChanged) {
                    this.beforeValueChanged(this, outerArgs);
                }
                if (this._valueChanged) {
                    this._valueChanged(this, outerArgs);
                }
            };
            this.i.valueChanged = delegateCombine(this.i.valueChanged, this._valueChanged_wrapped);
            ;
        }
        get actualMinimumValueChange() {
            return this._actualMinimumValueChange;
        }
        set actualMinimumValueChange(ev) {
            if (this._actualMinimumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
                this._actualMinimumValueChange_wrapped = null;
                this._actualMinimumValueChange = null;
            }
            this._actualMinimumValueChange = ev;
            this._actualMinimumValueChange_wrapped = (o, e) => {
                let ext = this.actualMinimumValue;
                if (e.propertyName == 'Bs') {
                    if (this.beforeActualMinimumValueChange) {
                        this.beforeActualMinimumValueChange(this, ext);
                    }
                    if (this._actualMinimumValueChange) {
                        this._actualMinimumValueChange(this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
        }
        get actualMaximumValueChange() {
            return this._actualMaximumValueChange;
        }
        set actualMaximumValueChange(ev) {
            if (this._actualMaximumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
                this._actualMaximumValueChange_wrapped = null;
                this._actualMaximumValueChange = null;
            }
            this._actualMaximumValueChange = ev;
            this._actualMaximumValueChange_wrapped = (o, e) => {
                let ext = this.actualMaximumValue;
                if (e.propertyName == 'Br') {
                    if (this.beforeActualMaximumValueChange) {
                        this.beforeActualMaximumValueChange(this, ext);
                    }
                    if (this._actualMaximumValueChange) {
                        this._actualMaximumValueChange(this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
        }
    }
    IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent = null;
    IgcRadialGaugeComponent.htmlTagName = "igc-radial-gauge";
    IgcRadialGaugeComponent._isElementRegistered = false;
    return IgcRadialGaugeComponent;
})();
