// igniteui-webcomponents-gauges/esm2015/lib/igc-linear-graph-range-component.js
import { TypeRegistrar } from "igniteui-webcomponents-core";
import { brushToString, stringToBrush, toSpinal, initializePropertiesFromCss, NamePatcher, getAllPropertyNames, fromSpinal } from "igniteui-webcomponents-core";
import { XamLinearGraphRange } from "./XamLinearGraphRange";
import { RegisterElementHelper } from "igniteui-webcomponents-core";
import { IgcHTMLElement } from "igniteui-webcomponents-core";
export let IgcLinearGraphRangeComponent = /*@__PURE__*/ (() => {
    class IgcLinearGraphRangeComponent extends IgcHTMLElement {
        constructor() {
            super();
            this._renderer = null;
            this._settingAttributes = false;
            this._attached = false;
            this._queuedSetAttributes = [];
            this._updatingFromAttribute = false;
            this.__p = null;
            this._hasUserValues = new Set();
            this._stylingContainer = null;
            this._stylingParent = null;
            this._inStyling = false;
            if (this._styling) {
                NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
            }
            this._implementation = this.createImplementation();
            this._implementation.externalObject = this;
            this.onImplementationCreated();
        }
        get i() {
            return this._implementation;
        }
        onImplementationCreated() {
        }
        get rangeInternal() {
            return this.i;
        }
        createImplementation() {
            return new XamLinearGraphRange();
        }
        ngOnInit() {
        }
        _provideRenderer(renderer) {
            this._renderer = renderer;
        }
        static _createFromInternal(internal) {
            if (!internal) {
                return null;
            }
            if (!internal.$type) {
                return null;
            }
            let name = internal.$type.name;
            let externalName = "Igc" + name + "Component";
            if (!TypeRegistrar.isRegistered(externalName)) {
                return null;
            }
            return TypeRegistrar.create(externalName);
        }
        _enqueueSetAttribute(attrName, attrValue) {
            this._queuedSetAttributes.push({ attrName: attrName, attrValue: attrValue });
        }
        _flushQueuedAttributes() {
            this._settingAttributes = true;
            for (let i = 0; i < this._queuedSetAttributes.length; i++) {
                this.setAttribute(this._queuedSetAttributes[i].attrName, this._queuedSetAttributes[i].attrValue);
            }
            this._settingAttributes = false;
            this._queuedSetAttributes.length = 0;
        }
        _a(attrName, attrValue) {
            if (this._updatingFromAttribute) {
                return;
            }
            if (attrValue) {
                attrValue = attrValue.toString();
            }
            this._settingAttributes = true;
            attrName = toSpinal(attrName);
            if (this._attached) {
                this.setAttribute(attrName, attrValue);
            }
            else {
                this._enqueueSetAttribute(attrName, attrValue);
            }
            this._settingAttributes = false;
        }
        static get observedAttributes() {
            if (IgcLinearGraphRangeComponent._observedAttributesIgcLinearGraphRangeComponent == null) {
                let names = getAllPropertyNames(IgcLinearGraphRangeComponent);
                for (let i = 0; i < names.length; i++) {
                    names[i] = toSpinal(names[i]);
                }
                IgcLinearGraphRangeComponent._observedAttributesIgcLinearGraphRangeComponent = names;
            }
            return IgcLinearGraphRangeComponent._observedAttributesIgcLinearGraphRangeComponent;
        }
        attributeChangedCallback(name, oldValue, newValue) {
            if (this._settingAttributes) {
                return;
            }
            let setName = fromSpinal(name);
            this._updatingFromAttribute = true;
            this[setName] = newValue;
            this._updatingFromAttribute = false;
        }
        static register() {
            if (!IgcLinearGraphRangeComponent._isElementRegistered) {
                IgcLinearGraphRangeComponent._isElementRegistered = true;
                RegisterElementHelper.registerElement(IgcLinearGraphRangeComponent.htmlTagName, IgcLinearGraphRangeComponent);
            }
        }
        /**
     * Gets or sets the name of the range.
    */
        get name() {
            return this.i.u;
        }
        set name(v) {
            this.i.u = v;
        }
        /**
         * Gets or sets the brush to use to fill the range.
        */
        get brush() {
            return brushToString(this.i.an);
        }
        set brush(v) {
            this.i.an = stringToBrush(v);
            this._a("brush", brushToString(this.i.an));
        }
        /**
         * Gets or sets the outline to use when rendering the range.
        */
        get outline() {
            return brushToString(this.i.ao);
        }
        set outline(v) {
            this.i.ao = stringToBrush(v);
            this._a("outline", brushToString(this.i.ao));
        }
        /**
         * Gets or sets the value at which the range starts along the scale.
        */
        get startValue() {
            return this.i.o;
        }
        set startValue(v) {
            this.i.o = +v;
            this._a("startValue", this.i.o);
        }
        /**
         * Gets or sets the value at which the range ends along the scale.
        */
        get endValue() {
            return this.i.j;
        }
        set endValue(v) {
            this.i.j = +v;
            this._a("endValue", this.i.j);
        }
        /**
         * Gets or sets the distance measured from the front/bottom of the bullet graph (from 0 to 1) at which to start rendering the inner edge of the range.
         * Values further from zero than 1 can be used to make this extend further than the normal width/height of the bullet graph.
        */
        get innerStartExtent() {
            return this.i.l;
        }
        set innerStartExtent(v) {
            this.i.l = +v;
            this._a("innerStartExtent", this.i.l);
        }
        /**
         * Gets or sets the distance measured from the front/bottom of the bullet graph (from 0 to 1) at which to end rendering the inner edge of the range.
         * Values further from zero than 1 can be used to make this extend further than the normal width/height of the bullet graph.
        */
        get innerEndExtent() {
            return this.i.k;
        }
        set innerEndExtent(v) {
            this.i.k = +v;
            this._a("innerEndExtent", this.i.k);
        }
        /**
         * Gets or sets the distance measured from the front/bottom of the bullet graph (from 0 to 1) at which to start rendering the outer edge of the range.
         * Values further from zero than 1 can be used to make this extend further than the normal width/height of the bullet graph.
        */
        get outerStartExtent() {
            return this.i.n;
        }
        set outerStartExtent(v) {
            this.i.n = +v;
            this._a("outerStartExtent", this.i.n);
        }
        /**
         * Gets or sets the distance measured from the front/bottom of the bullet graph (from 0 to 1) at which to end rendering the outer edge of the range.
         * Values further from zero than 1 can be used to make this extend further than the normal width/height of the bullet graph.
        */
        get outerEndExtent() {
            return this.i.m;
        }
        set outerEndExtent(v) {
            this.i.m = +v;
            this._a("outerEndExtent", this.i.m);
        }
        /**
         * Gets or sets the stroke thickness to use when rendering this range's outline.
        */
        get strokeThickness() {
            return this.i.p;
        }
        set strokeThickness(v) {
            this.i.p = +v;
            this._a("strokeThickness", this.i.p);
        }
        findByName(name) {
            if (this.findEphemera) {
                if (name && name.indexOf("@@e:") == 0) {
                    return this.findEphemera(name);
                }
            }
            return null;
        }
        get hasUserValues() {
            return this._hasUserValues;
        }
        __m(propertyName) {
            if (!this._inStyling) {
                this._hasUserValues.add(propertyName);
            }
        }
        _styling(container, component, parent) {
            if (this._inStyling) {
                return;
            }
            this._inStyling = true;
            this._stylingContainer = container;
            this._stylingParent = component;
            let genericPrefix = "";
            let typeName = this.i.$type.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            genericPrefix = toSpinal("LinearGraphRangeComponent");
            let additionalPrefixes = [];
            let prefix = toSpinal(typeName);
            additionalPrefixes.push(prefix + "-");
            let b = this.i.$type.baseType;
            while (b && b.name != "Object" &&
                b.name != "Base" &&
                b.name != "Control" &&
                b.Name != "DependencyObject" &&
                b.Name != "FrameworkElement") {
                typeName = b.name;
                if (typeName.indexOf("Xam") === 0) {
                    typeName = typeName.substring(3);
                }
                let basePrefix = toSpinal(typeName);
                additionalPrefixes.push(basePrefix + "-");
                b = b.baseType;
            }
            if (parent) {
                let parentTypeName = parent.i.$type.name;
                if (parentTypeName.indexOf("Xam") === 0) {
                    parentTypeName = parentTypeName.substring(3);
                }
                let parentPrefix = toSpinal(parentTypeName);
                additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
                additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
            }
            initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
            if (this._otherStyling) {
                this._otherStyling(container, component, parent);
            }
            this._inStyling = false;
        }
    }
    IgcLinearGraphRangeComponent._observedAttributesIgcLinearGraphRangeComponent = null;
    IgcLinearGraphRangeComponent.htmlTagName = "igc-linear-graph-range";
    IgcLinearGraphRangeComponent._isElementRegistered = false;
    return IgcLinearGraphRangeComponent;
})();
