// igniteui-webcomponents-core/esm2015/lib/WebExporterOptionsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebExporterOptionsBaseDescription = /*@__PURE__*/ (() => {
    class WebExporterOptionsBaseDescription extends Description {
        constructor() {
            super();
            this.i = false;
            this.j = false;
            this.h = false;
            this.m = false;
            this.k = false;
            this.l = false;
            this.g = false;
            this.f = false;
            this.v = null;
        }
        get_type() {
            return "WebExporterOptionsBase";
        }
        get type() {
            return this.get_type();
        }
        get ignoreColumnsVisibility() {
            return this.i;
        }
        set ignoreColumnsVisibility(a) {
            this.i = a;
            this.e("IgnoreColumnsVisibility");
        }
        get ignoreFiltering() {
            return this.j;
        }
        set ignoreFiltering(a) {
            this.j = a;
            this.e("IgnoreFiltering");
        }
        get ignoreColumnsOrder() {
            return this.h;
        }
        set ignoreColumnsOrder(a) {
            this.h = a;
            this.e("IgnoreColumnsOrder");
        }
        get ignoreSorting() {
            return this.m;
        }
        set ignoreSorting(a) {
            this.m = a;
            this.e("IgnoreSorting");
        }
        get ignoreGrouping() {
            return this.k;
        }
        set ignoreGrouping(a) {
            this.k = a;
            this.e("IgnoreGrouping");
        }
        get ignoreMultiColumnHeaders() {
            return this.l;
        }
        set ignoreMultiColumnHeaders(a) {
            this.l = a;
            this.e("IgnoreMultiColumnHeaders");
        }
        get freezeHeaders() {
            return this.g;
        }
        set freezeHeaders(a) {
            this.g = a;
            this.e("FreezeHeaders");
        }
        get alwaysExportHeaders() {
            return this.f;
        }
        set alwaysExportHeaders(a) {
            this.f = a;
            this.e("AlwaysExportHeaders");
        }
        get fileName() {
            return this.v;
        }
        set fileName(a) {
            this.v = a;
            this.e("FileName");
        }
    }
    WebExporterOptionsBaseDescription.$t = /*@__PURE__*/ markType(WebExporterOptionsBaseDescription, 'WebExporterOptionsBaseDescription', Description.$);
    return WebExporterOptionsBaseDescription;
})();
