// igniteui-webcomponents-core/esm2015/lib/WebRowSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowSelectionEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebRowSelectionEventArgsDetailDescription extends WebBaseEventArgsDetailDescription {
        constructor() {
            super();
            this.o = null;
            this.n = null;
            this.m = null;
            this.p = null;
        }
        get_type() {
            return "WebRowSelectionEventArgsDetail";
        }
        get oldSelectionRef() {
            return this.o;
        }
        set oldSelectionRef(a) {
            this.o = a;
            this.e("OldSelectionRef");
        }
        get newSelectionRef() {
            return this.n;
        }
        set newSelectionRef(a) {
            this.n = a;
            this.e("NewSelectionRef");
        }
        get addedRef() {
            return this.m;
        }
        set addedRef(a) {
            this.m = a;
            this.e("AddedRef");
        }
        get removedRef() {
            return this.p;
        }
        set removedRef(a) {
            this.p = a;
            this.e("RemovedRef");
        }
    }
    WebRowSelectionEventArgsDetailDescription.$t = /*@__PURE__*/ markType(WebRowSelectionEventArgsDetailDescription, 'WebRowSelectionEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowSelectionEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowSelectionEventArgsDetailDescription.__marshalByValueAlias1 = "RowSelectionEventArgsDetail";
    return WebRowSelectionEventArgsDetailDescription;
})();
