// igniteui-webcomponents-core/esm5/lib/CalloutRenderStyleUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutRenderStyleUpdatingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(CalloutRenderStyleUpdatingEventArgsDescription, _super);
    function CalloutRenderStyleUpdatingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.x = null;
        _this.y = null;
        _this.h = 0;
        _this.i = 0;
        _this.k = 0;
        _this.l = 0;
        _this.r = null;
        _this.t = null;
        _this.f = null;
        _this.s = null;
        _this.w = null;
        _this.v = null;
        _this.u = null;
        _this.j = 0;
        return _this;
    }
    CalloutRenderStyleUpdatingEventArgsDescription.prototype.get_type = function () {
        return "CalloutRenderStyleUpdatingEventArgs";
    };
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.e("XValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.e("YValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "labelPositionX", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("LabelPositionX");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "labelPositionY", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("LabelPositionY");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "targetPositionX", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("TargetPositionX");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "targetPositionY", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("TargetPositionY");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "actualPosition", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("ActualPosition");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("ItemRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "series", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Series");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "background", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("Background");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "textColor", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.e("TextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "outline", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("Outline");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "leaderBrush", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.e("LeaderBrush");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutRenderStyleUpdatingEventArgsDescription.prototype, "strokeThickness", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("StrokeThickness");
        },
        enumerable: true,
        configurable: true
    });
    CalloutRenderStyleUpdatingEventArgsDescription.$t = markType(CalloutRenderStyleUpdatingEventArgsDescription, 'CalloutRenderStyleUpdatingEventArgsDescription', Description.$);
    return CalloutRenderStyleUpdatingEventArgsDescription;
}(Description));
export { CalloutRenderStyleUpdatingEventArgsDescription };
