// igniteui-webcomponents-core/esm5/lib/CodeGeneratingComponentRendererModuleEmitter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, typeCast, markType } from "./type";
import { CodeGeneratingCodeWriter } from "./CodeGeneratingCodeWriter";
import { CodeGeneratingImportManager } from "./CodeGeneratingImportManager";
import { JsonDictionaryArray } from "./JsonDictionaryArray";
import { JsonDictionaryValue } from "./JsonDictionaryValue";
import { stringStartsWith, stringReplace } from "./string";
/**
 * @hidden
 */
var CodeGeneratingComponentRendererModuleEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(CodeGeneratingComponentRendererModuleEmitter, _super);
    function CodeGeneratingComponentRendererModuleEmitter(a, b) {
        var _this = _super.call(this) || this;
        _this.k = null;
        _this.i = null;
        _this.b = null;
        _this.f = new CodeGeneratingImportManager();
        _this.e = new CodeGeneratingImportManager();
        _this.a = null;
        _this.k = a;
        _this.i = b;
        _this.b = new CodeGeneratingCodeWriter(_this.k);
        _this.a = new CodeGeneratingCodeWriter(_this.k);
        return _this;
    }
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "j", {
        get: function () {
            return this.i;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "d", {
        get: function () {
            return this.b;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "h", {
        get: function () {
            return this.f;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "g", {
        get: function () {
            return this.e;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeGeneratingComponentRendererModuleEmitter.prototype, "c", {
        get: function () {
            return this.a;
        },
        enumerable: true,
        configurable: true
    });
    CodeGeneratingComponentRendererModuleEmitter.prototype.n = function (a, b) {
        this.p();
        if (typeCast(JsonDictionaryArray.$, a) !== null) {
            var c = a;
            for (var d = 0; d < c.items.length; d++) {
                var e = c.items[d];
                if (typeCast(JsonDictionaryValue.$, e) !== null) {
                    var f = e;
                    if (f != null) {
                        var g = f.value.toString();
                        g = this.l(g, b);
                        this.m(g);
                    }
                }
            }
        }
        this.o();
        this.i.addImportResult("modulesImports", this.h);
        this.i.addContainerResult("modulesRegister", this.d.toString());
        this.i.addImportResult("descriptionImports", this.g);
        this.i.addContainerResult("descriptionRegister", this.c.toString());
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.l = function (a, b) {
        if (a == "withDescriptions") {
            return a;
        }
        var c = "";
        var d = a.split('/');
        if (d.length == 1) {
            c = d[0];
            return c;
        }
        else {
            var e = d[0];
            c = d[1];
            if (stringStartsWith(c, "Web") && b.getMetadata(stringReplace(c, "Module", ""), "__qualifiedNameTS") != null) {
                c = c.substr(3);
                this.j.addPackageRef("web" + e);
            }
            return e + "/" + c;
        }
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.o = function () {
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.m = function (a) {
    };
    CodeGeneratingComponentRendererModuleEmitter.prototype.p = function () {
    };
    CodeGeneratingComponentRendererModuleEmitter.$t = markType(CodeGeneratingComponentRendererModuleEmitter, 'CodeGeneratingComponentRendererModuleEmitter');
    return CodeGeneratingComponentRendererModuleEmitter;
}(Base));
export { CodeGeneratingComponentRendererModuleEmitter };
