// igniteui-webcomponents-core/esm5/lib/DataGridPagerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridPagerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DataGridPagerDescription, _super);
    function DataGridPagerDescription() {
        var _this = _super.call(this) || this;
        _this.w = null;
        _this.l = 0;
        _this.k = 0;
        _this.r = null;
        _this.f = null;
        _this.j = 0;
        _this.s = null;
        _this.h = 0;
        _this.t = null;
        _this.u = null;
        _this.p = null;
        _this.q = null;
        _this.v = null;
        return _this;
    }
    DataGridPagerDescription.prototype.get_type = function () {
        return "DataGridPager";
    };
    Object.defineProperty(DataGridPagerDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pagerText", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.e("PagerText");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageSize", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("PageSize");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageCount", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("PageCount");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("DataSourceRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "dataSource", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("DataSource");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "currentPage", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("CurrentPage");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontFamily", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("FontFamily");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontSize", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("FontSize");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontStyle", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("FontStyle");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "fontWeight", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.e("FontWeight");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "background", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("Background");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "border", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("Border");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridPagerDescription.prototype, "pageChangedRef", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("PageChangedRef");
        },
        enumerable: true,
        configurable: true
    });
    DataGridPagerDescription.$t = markType(DataGridPagerDescription, 'DataGridPagerDescription', Description.$);
    return DataGridPagerDescription;
}(Description));
export { DataGridPagerDescription };
