// igniteui-webcomponents-core/esm5/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicPolylineSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(GeographicPolylineSeriesDescription, _super);
    function GeographicPolylineSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.dz = null;
        _this.dx = null;
        _this.dy = null;
        _this.du = 0;
        _this.dt = 0;
        _this.d0 = null;
        return _this;
    }
    GeographicPolylineSeriesDescription.prototype.get_type = function () {
        return "GeographicPolylineSeries";
    };
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStyleSelectorRef", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.e("ShapeStyleSelectorRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeFill", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.e("ShapeFill");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStroke", {
        get: function () {
            return this.dy;
        },
        set: function (a) {
            this.dy = a;
            this.e("ShapeStroke");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.du;
        },
        set: function (a) {
            this.du = a;
            this.e("ShapeStrokeThickness");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.dt;
        },
        set: function (a) {
            this.dt = a;
            this.e("ShapeOpacity");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.e("StyleShapeRef");
        },
        enumerable: true,
        configurable: true
    });
    GeographicPolylineSeriesDescription.$t = markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
}(GeographicShapeSeriesBaseDescription));
export { GeographicPolylineSeriesDescription };
