// igniteui-webcomponents-core/esm5/lib/HorizontalRangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { RangeCategorySeriesDescription } from "./RangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HorizontalRangeCategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(HorizontalRangeCategorySeriesDescription, _super);
    function HorizontalRangeCategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.eb = null;
        _this.ec = null;
        _this.d9 = null;
        _this.d7 = null;
        _this.ea = null;
        _this.d8 = null;
        return _this;
    }
    HorizontalRangeCategorySeriesDescription.prototype.get_type = function () {
        return "HorizontalRangeCategorySeries";
    };
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.e("XAxisRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.e("YAxisRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "lowMemberAsLegendLabel", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.e("LowMemberAsLegendLabel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "highMemberAsLegendLabel", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.e("HighMemberAsLegendLabel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "lowMemberAsLegendUnit", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.e("LowMemberAsLegendUnit");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HorizontalRangeCategorySeriesDescription.prototype, "highMemberAsLegendUnit", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.e("HighMemberAsLegendUnit");
        },
        enumerable: true,
        configurable: true
    });
    HorizontalRangeCategorySeriesDescription.$t = markType(HorizontalRangeCategorySeriesDescription, 'HorizontalRangeCategorySeriesDescription', RangeCategorySeriesDescription.$);
    return HorizontalRangeCategorySeriesDescription;
}(RangeCategorySeriesDescription));
export { HorizontalRangeCategorySeriesDescription };
