// igniteui-webcomponents-core/esm5/lib/igc-data-legend-series-context.js
import { DataLegendSeriesGroup_$type } from "./DataLegendSeriesGroup";
import { IgcDataLegendSeriesValueInfo } from "./igc-data-legend-series-value-info";
import { DataLegendSeriesContext as DataLegendSeriesContext_internal } from "./DataLegendSeriesContext";
import { ensureEnum } from "./componentUtil";
/**
 * Represent info about series values
*/
var IgcDataLegendSeriesContext = /** @class */ /*@__PURE__*/ (function () {
    function IgcDataLegendSeriesContext() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    IgcDataLegendSeriesContext.prototype.createImplementation = function () {
        return new DataLegendSeriesContext_internal();
    };
    Object.defineProperty(IgcDataLegendSeriesContext.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: true,
        configurable: true
    });
    IgcDataLegendSeriesContext.prototype.onImplementationCreated = function () {
    };
    IgcDataLegendSeriesContext.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    };
    Object.defineProperty(IgcDataLegendSeriesContext.prototype, "seriesGroup", {
        /**
         * Gets a group of series that the current series belongs to
        */
        get: function () {
            return this.i.b;
        },
        set: function (v) {
            this.i.b = ensureEnum(DataLegendSeriesGroup_$type, v);
        },
        enumerable: true,
        configurable: true
    });
    IgcDataLegendSeriesContext.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    /**
     * Creates an instance of DataLegendSeriesContext
    
    */
    IgcDataLegendSeriesContext.prototype.getSeriesValueInfo = function (type) {
        var iv = this.i.c((type == null ? null : type));
        var ret = null;
        if (iv && iv.externalObject) {
            ret = iv.externalObject;
        }
        else {
            if (iv) {
                var e = new IgcDataLegendSeriesValueInfo();
                e._implementation = iv;
                iv.externalObject = e;
                ret = e;
            }
        }
        return ret;
    };
    /**
     * Gets value info for the series value type
    
    */
    IgcDataLegendSeriesContext.prototype.getSeriesValue = function (type) {
        var iv = this.i.g((type == null ? null : type));
        return (iv);
    };
    IgcDataLegendSeriesContext.prototype.setSeriesValue = function (type, value) {
        this.i.k((type == null ? null : type), value);
    };
    IgcDataLegendSeriesContext.prototype.setSeriesValueInfo = function (type, valueInfo) {
        this.i.l((type == null ? null : type), (valueInfo == null ? null : valueInfo.i));
    };
    /**
     * Gets all values that a series renders at the current data point
     * For example, Open, High, Low, Close for Financial Series
    
    */
    IgcDataLegendSeriesContext.prototype.getSeriesValues = function () {
        var iv = this.i.a();
        return (iv);
    };
    return IgcDataLegendSeriesContext;
}());
export { IgcDataLegendSeriesContext };
