// igniteui-webcomponents-core/esm5/lib/PolarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var PolarBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(PolarBaseDescription, _super);
    function PolarBaseDescription() {
        var _this = _super.call(this) || this;
        _this.ec = null;
        _this.ej = null;
        _this.d9 = null;
        _this.eg = null;
        _this.ds = false;
        _this.d0 = 0;
        _this.el = null;
        _this.ek = null;
        _this.d8 = null;
        _this.dx = 0;
        _this.d1 = 0;
        _this.d2 = 0;
        _this.dp = false;
        _this.d7 = null;
        _this.ef = null;
        _this.dz = 0;
        _this.dr = false;
        _this.dq = false;
        _this.eh = null;
        _this.ea = null;
        _this.ei = null;
        _this.eb = null;
        _this.ee = null;
        _this.ed = null;
        return _this;
    }
    PolarBaseDescription.prototype.get_type = function () {
        return "PolarBase";
    };
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberPath", {
        get: function () {
            return this.ec;
        },
        set: function (a) {
            this.ec = a;
            this.e("AngleMemberPath");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.ej;
        },
        set: function (a) {
            this.ej = a;
            this.e("RadiusMemberPath");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleAxisRef", {
        get: function () {
            return this.d9;
        },
        set: function (a) {
            this.d9 = a;
            this.e("AngleAxisRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusAxisRef", {
        get: function () {
            return this.eg;
        },
        set: function (a) {
            this.eg = a;
            this.e("RadiusAxisRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "useCartesianInterpolation", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.e("UseCartesianInterpolation");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "maximumMarkers", {
        get: function () {
            return this.d0;
        },
        set: function (a) {
            this.d0 = a;
            this.e("MaximumMarkers");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineType", {
        get: function () {
            return this.el;
        },
        set: function (a) {
            this.el = a;
            this.e("TrendLineType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.ek;
        },
        set: function (a) {
            this.ek = a;
            this.e("TrendLineBrush");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.d8;
        },
        set: function (a) {
            this.d8 = a;
            this.e("ActualTrendLineBrush");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.dx;
        },
        set: function (a) {
            this.dx = a;
            this.e("TrendLineThickness");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.d1;
        },
        set: function (a) {
            this.d1 = a;
            this.e("TrendLinePeriod");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.d2;
        },
        set: function (a) {
            this.d2 = a;
            this.e("TrendLineZIndex");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "clipSeriesToBounds", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.e("ClipSeriesToBounds");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.d7;
        },
        set: function (a) {
            this.d7 = a;
            this.e("ActualItemSearchMode");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.ef;
        },
        set: function (a) {
            this.ef = a;
            this.e("ItemSearchMode");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.dz;
        },
        set: function (a) {
            this.dz = a;
            this.e("ItemSearchThreshold");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarStyleAllowed", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.e("IsCustomPolarStyleAllowed");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarMarkerStyleAllowed", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.e("IsCustomPolarMarkerStyleAllowed");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendLabel", {
        get: function () {
            return this.eh;
        },
        set: function (a) {
            this.eh = a;
            this.e("RadiusMemberAsLegendLabel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendLabel", {
        get: function () {
            return this.ea;
        },
        set: function (a) {
            this.ea = a;
            this.e("AngleMemberAsLegendLabel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendUnit", {
        get: function () {
            return this.ei;
        },
        set: function (a) {
            this.ei = a;
            this.e("RadiusMemberAsLegendUnit");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendUnit", {
        get: function () {
            return this.eb;
        },
        set: function (a) {
            this.eb = a;
            this.e("AngleMemberAsLegendUnit");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarStyleRef", {
        get: function () {
            return this.ee;
        },
        set: function (a) {
            this.ee = a;
            this.e("AssigningPolarStyleRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarMarkerStyleRef", {
        get: function () {
            return this.ed;
        },
        set: function (a) {
            this.ed = a;
            this.e("AssigningPolarMarkerStyleRef");
        },
        enumerable: true,
        configurable: true
    });
    PolarBaseDescription.$t = markType(PolarBaseDescription, 'PolarBaseDescription', MarkerSeriesDescription.$);
    return PolarBaseDescription;
}(MarkerSeriesDescription));
export { PolarBaseDescription };
