// igniteui-webcomponents-core/esm5/lib/PropertyEditorDescriptionObjectDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var PropertyEditorDescriptionObjectDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(PropertyEditorDescriptionObjectDescription, _super);
    function PropertyEditorDescriptionObjectDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.f = null;
        return _this;
    }
    PropertyEditorDescriptionObjectDescription.prototype.get_type = function () {
        return "PropertyEditorDescriptionObject";
    };
    Object.defineProperty(PropertyEditorDescriptionObjectDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorDescriptionObjectDescription.prototype, "descriptionType", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("DescriptionType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorDescriptionObjectDescription.prototype, "properties", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Properties");
        },
        enumerable: true,
        configurable: true
    });
    PropertyEditorDescriptionObjectDescription.$t = markType(PropertyEditorDescriptionObjectDescription, 'PropertyEditorDescriptionObjectDescription', Description.$);
    return PropertyEditorDescriptionObjectDescription;
}(Description));
export { PropertyEditorDescriptionObjectDescription };
/**
 * @hidden
 */
var PropertyEditorPropertyDescriptionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(PropertyEditorPropertyDescriptionDescription, _super);
    function PropertyEditorPropertyDescriptionDescription() {
        var _this = _super.call(this) || this;
        _this.ag = null;
        _this.ae = null;
        _this.ai = null;
        _this.af = null;
        _this.n = false;
        _this.ad = null;
        _this.g = null;
        _this.h = null;
        _this.z = null;
        _this.s = 0;
        _this.p = 0;
        _this.q = 0;
        _this.r = 0;
        _this.t = 0;
        _this.l = null;
        _this.f = null;
        _this.ab = null;
        _this.ac = null;
        _this.ah = null;
        return _this;
    }
    PropertyEditorPropertyDescriptionDescription.prototype.get_type = function () {
        return "PropertyEditorPropertyDescription";
    };
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "propertyPath", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.e("PropertyPath");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "label", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.e("Label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "valueType", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.e("ValueType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "propertyDescriptionType", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.e("PropertyDescriptionType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "shouldOverrideDefaultEditor", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("ShouldOverrideDefaultEditor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "elementDescriptionType", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.e("ElementDescriptionType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "dropDownNames", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("DropDownNames");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "dropDownValues", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("DropDownValues");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "primitiveValue", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.e("PrimitiveValue");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "min", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("Min");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "editorWidth", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("EditorWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "labelWidth", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("LabelWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "max", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("Max");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "step", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("Step");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "complexValue", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("ComplexValue");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "complexValues", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("ComplexValues");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "buttonClickedRef", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.e("ButtonClickedRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "changedRef", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.e("ChangedRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PropertyEditorPropertyDescriptionDescription.prototype, "targetPropertyUpdatingRef", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.e("TargetPropertyUpdatingRef");
        },
        enumerable: true,
        configurable: true
    });
    PropertyEditorPropertyDescriptionDescription.$t = markType(PropertyEditorPropertyDescriptionDescription, 'PropertyEditorPropertyDescriptionDescription', Description.$);
    return PropertyEditorPropertyDescriptionDescription;
}(Description));
export { PropertyEditorPropertyDescriptionDescription };
