// igniteui-webcomponents-core/esm5/lib/TSCodeGeneratingComponentRendererModuleEmitter.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { CodeGeneratingComponentRendererModuleEmitter } from "./CodeGeneratingComponentRendererModuleEmitter";
import { Dictionary$2 } from "./Dictionary$2";
import { List$1 } from "./List$1";
import { String_$type, fromEnum, markType } from "./type";
import { stringEndsWith } from "./string";
/**
 * @hidden
 */
var TSCodeGeneratingComponentRendererModuleEmitter = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(TSCodeGeneratingComponentRendererModuleEmitter, _super);
    function TSCodeGeneratingComponentRendererModuleEmitter(a, b, c) {
        var _this = _super.call(this, b, c) || this;
        _this.q = 0;
        _this.t = 0;
        _this.r = false;
        _this.s = new Dictionary$2(String_$type, List$1.$.specialize(String_$type), 0);
        _this.q = a;
        return _this;
    }
    TSCodeGeneratingComponentRendererModuleEmitter.prototype.p = function () {
        _super.prototype.p.call(this);
        this.t = 0;
    };
    Object.defineProperty(TSCodeGeneratingComponentRendererModuleEmitter.prototype, "v", {
        get: function () {
            switch (this.q) {
                case 1: return "Igx";
                case 2: return "Igr";
                case 3: return "Igc";
            }
            return "";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TSCodeGeneratingComponentRendererModuleEmitter.prototype, "u", {
        get: function () {
            switch (this.q) {
                case 1: return "angular";
                case 2: return "react";
                case 3: return "webcomponents";
            }
            return "";
        },
        enumerable: true,
        configurable: true
    });
    TSCodeGeneratingComponentRendererModuleEmitter.prototype.m = function (a) {
        _super.prototype.m.call(this, a);
        if (a == "withDescriptions") {
            this.r = true;
            return;
        }
        var b = "igniteui-" + this.u + "-core";
        var c = "";
        var d = a.split('/');
        if (d.length == 1) {
            c = d[0];
            this.j.addPackageRef(c);
        }
        else {
            b = d[0];
            this.j.addPackageRef(b);
            b = "igniteui-" + this.u + "-" + b;
            c = d[1];
        }
        if (!this.s.containsKey(b)) {
            this.s.addItem(b, new List$1(String_$type, 0));
        }
        var e = this.s.item(b);
        if (this.t > 0) {
            this.d.l(",");
        }
        this.d.j(this.v + c);
        e.add(this.v + c);
        if (this.r) {
            if (this.t == 0) {
                this.c.l("private _componentRenderer: ComponentRenderer = null;");
                this.c.l("public get renderer(): ComponentRenderer {");
                this.c.h();
                this.c.l("if (this._componentRenderer == null) {");
                this.c.h();
                this.c.l("this._componentRenderer = new ComponentRenderer();");
                this.c.l("var context = this._componentRenderer.context;");
                this.g.addPackageImport("igniteui-" + this.u + "-core", "ComponentRenderer");
            }
            var f = c;
            if (stringEndsWith(f, "Module")) {
                f = f.substr(0, f.length - ("Module").length);
                f += "DescriptionModule";
            }
            this.g.addPackageImport("igniteui-" + this.u + "-core", f);
            this.c.l(f + ".register(context);");
        }
        this.t++;
    };
    TSCodeGeneratingComponentRendererModuleEmitter.prototype.o = function () {
        var e_1, _a, e_2, _b;
        _super.prototype.o.call(this);
        try {
            for (var _c = tslib_1.__values(fromEnum(this.s.keys)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var a = _d.value;
                var b = a;
                var c = this.s.item(a);
                try {
                    for (var _e = (e_2 = void 0, tslib_1.__values(fromEnum(c))), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var d = _f.value;
                        this.h.addPackageImport(b, d);
                    }
                }
                catch (e_2_1) {
                    e_2 = { error: e_2_1 };
                }
                finally {
                    try {
                        if (_f && !_f.done && (_b = _e.return))
                            _b.call(_e);
                    }
                    finally {
                        if (e_2)
                            throw e_2.error;
                    }
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        if (this.r) {
            this.c.f();
            this.c.l("}");
            this.c.l("return this._componentRenderer;");
            this.c.f();
            this.c.l("}");
        }
    };
    TSCodeGeneratingComponentRendererModuleEmitter.$t = markType(TSCodeGeneratingComponentRendererModuleEmitter, 'TSCodeGeneratingComponentRendererModuleEmitter', CodeGeneratingComponentRendererModuleEmitter.$);
    return TSCodeGeneratingComponentRendererModuleEmitter;
}(CodeGeneratingComponentRendererModuleEmitter));
export { TSCodeGeneratingComponentRendererModuleEmitter };
