// igniteui-webcomponents-core/esm5/lib/WebColumnExportingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnExportingEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebColumnExportingEventArgsDetailDescription, _super);
    function WebColumnExportingEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.v = null;
        _this.u = null;
        _this.r = 0;
        _this.n = false;
        _this.o = false;
        _this.k = null;
        return _this;
    }
    WebColumnExportingEventArgsDetailDescription.prototype.get_type = function () {
        return "WebColumnExportingEventArgsDetail";
    };
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "header", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("Header");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "field", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.e("Field");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "columnIndex", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("ColumnIndex");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "cancel", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("Cancel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "skipFormatter", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("SkipFormatter");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebColumnExportingEventArgsDetailDescription.prototype, "grid", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("Grid");
        },
        enumerable: true,
        configurable: true
    });
    WebColumnExportingEventArgsDetailDescription.$t = markType(WebColumnExportingEventArgsDetailDescription, 'WebColumnExportingEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebColumnExportingEventArgsDetailDescription.__marshalByValue1 = true;
    WebColumnExportingEventArgsDetailDescription.__marshalByValueAlias1 = "ColumnExportingEventArgsDetail";
    return WebColumnExportingEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebColumnExportingEventArgsDetailDescription };
