// igniteui-webcomponents-core/esm5/lib/WebExporterOptionsBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebExporterOptionsBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebExporterOptionsBaseDescription, _super);
    function WebExporterOptionsBaseDescription() {
        var _this = _super.call(this) || this;
        _this.i = false;
        _this.j = false;
        _this.h = false;
        _this.m = false;
        _this.k = false;
        _this.l = false;
        _this.g = false;
        _this.f = false;
        _this.v = null;
        return _this;
    }
    WebExporterOptionsBaseDescription.prototype.get_type = function () {
        return "WebExporterOptionsBase";
    };
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreColumnsVisibility", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("IgnoreColumnsVisibility");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreFiltering", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("IgnoreFiltering");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreColumnsOrder", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("IgnoreColumnsOrder");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreSorting", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("IgnoreSorting");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreGrouping", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("IgnoreGrouping");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "ignoreMultiColumnHeaders", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("IgnoreMultiColumnHeaders");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "freezeHeaders", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("FreezeHeaders");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "alwaysExportHeaders", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("AlwaysExportHeaders");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebExporterOptionsBaseDescription.prototype, "fileName", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("FileName");
        },
        enumerable: true,
        configurable: true
    });
    WebExporterOptionsBaseDescription.$t = markType(WebExporterOptionsBaseDescription, 'WebExporterOptionsBaseDescription', Description.$);
    return WebExporterOptionsBaseDescription;
}(Description));
export { WebExporterOptionsBaseDescription };
