// igniteui-webcomponents-core/esm5/lib/WebFloatPaneActionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { WebPaneDragActionDescription } from "./WebPaneDragActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebFloatPaneActionDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebFloatPaneActionDescription, _super);
    function WebFloatPaneActionDescription() {
        var _this = _super.call(this) || this;
        _this.m = null;
        _this.g = null;
        _this.j = 0;
        _this.i = 0;
        return _this;
    }
    WebFloatPaneActionDescription.prototype.get_type = function () {
        return "WebFloatPaneAction";
    };
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "actionType", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("ActionType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "location", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("Location");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "width", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("Width");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebFloatPaneActionDescription.prototype, "height", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Height");
        },
        enumerable: true,
        configurable: true
    });
    WebFloatPaneActionDescription.$t = markType(WebFloatPaneActionDescription, 'WebFloatPaneActionDescription', WebPaneDragActionDescription.$);
    return WebFloatPaneActionDescription;
}(WebPaneDragActionDescription));
export { WebFloatPaneActionDescription };
