// igniteui-webcomponents-core/esm5/lib/WebInputBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebInputBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebInputBaseDescription, _super);
    function WebInputBaseDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.g = false;
        _this.i = false;
        _this.f = false;
        _this.h = false;
        _this.s = null;
        _this.q = null;
        _this.t = null;
        _this.p = null;
        _this.o = null;
        _this.n = null;
        return _this;
    }
    WebInputBaseDescription.prototype.get_type = function () {
        return "WebInputBase";
    };
    Object.defineProperty(WebInputBaseDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "name", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("Name");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "outlined", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("Outlined");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "required", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Required");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "disabled", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Disabled");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "readonly", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("Readonly");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "placeholder", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("Placeholder");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "label", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("Label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "size", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("Size");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "inputOcurredRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("InputOcurredRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "focusRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("FocusRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebInputBaseDescription.prototype, "blurRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("BlurRef");
        },
        enumerable: true,
        configurable: true
    });
    WebInputBaseDescription.$t = markType(WebInputBaseDescription, 'WebInputBaseDescription', Description.$);
    return WebInputBaseDescription;
}(Description));
export { WebInputBaseDescription };
