// igniteui-webcomponents-core/esm5/lib/WebRowSelectionEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { WebBaseEventArgsDetailDescription } from "./WebBaseEventArgsDetailDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebRowSelectionEventArgsDetailDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebRowSelectionEventArgsDetailDescription, _super);
    function WebRowSelectionEventArgsDetailDescription() {
        var _this = _super.call(this) || this;
        _this.o = null;
        _this.n = null;
        _this.m = null;
        _this.p = null;
        return _this;
    }
    WebRowSelectionEventArgsDetailDescription.prototype.get_type = function () {
        return "WebRowSelectionEventArgsDetail";
    };
    Object.defineProperty(WebRowSelectionEventArgsDetailDescription.prototype, "oldSelectionRef", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("OldSelectionRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebRowSelectionEventArgsDetailDescription.prototype, "newSelectionRef", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("NewSelectionRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebRowSelectionEventArgsDetailDescription.prototype, "addedRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("AddedRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebRowSelectionEventArgsDetailDescription.prototype, "removedRef", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("RemovedRef");
        },
        enumerable: true,
        configurable: true
    });
    WebRowSelectionEventArgsDetailDescription.$t = markType(WebRowSelectionEventArgsDetailDescription, 'WebRowSelectionEventArgsDetailDescription', WebBaseEventArgsDetailDescription.$);
    WebRowSelectionEventArgsDetailDescription.__marshalByValue1 = true;
    WebRowSelectionEventArgsDetailDescription.__marshalByValueAlias1 = "RowSelectionEventArgsDetail";
    return WebRowSelectionEventArgsDetailDescription;
}(WebBaseEventArgsDetailDescription));
export { WebRowSelectionEventArgsDetailDescription };
