// igniteui-webcomponents-core/esm5/lib/WebSummaryResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebSummaryResultDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(WebSummaryResultDescription, _super);
    function WebSummaryResultDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.i = null;
        _this.g = false;
        return _this;
    }
    WebSummaryResultDescription.prototype.get_type = function () {
        return "WebSummaryResult";
    };
    Object.defineProperty(WebSummaryResultDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "key", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("Key");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "label", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("Label");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "result", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Result");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WebSummaryResultDescription.prototype, "defaultFormatting", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("DefaultFormatting");
        },
        enumerable: true,
        configurable: true
    });
    WebSummaryResultDescription.$t = markType(WebSummaryResultDescription, 'WebSummaryResultDescription', Description.$);
    WebSummaryResultDescription.__marshalByValue = true;
    WebSummaryResultDescription.__marshalByValueAlias = "SummaryResult";
    return WebSummaryResultDescription;
}(Description));
export { WebSummaryResultDescription };
