// igniteui-webcomponents-gauges/esm5/lib/igc-linear-gauge-component.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { WebComponentRenderer, WebComponentWrapper } from "igniteui-webcomponents-core";
import { TypeRegistrar } from "igniteui-webcomponents-core";
import { XamLinearGauge } from './XamLinearGauge';
import { CollectionAdapter, brushToString, stringToBrush, toPoint, fromBrushCollection, toBrushCollection, ensureBool, ensureEnum, initializePropertiesFromCss, NamePatcher, toSpinal, fromSpinal, getAllPropertyNames, enumToString, brushCollectionToString, arrayFindByName } from "igniteui-webcomponents-core";
import { IgcDataContext } from "igniteui-webcomponents-core";
import { delegateCombine, delegateRemove } from "igniteui-webcomponents-core";
import { NotifyCollectionChangedAction } from "igniteui-webcomponents-core";
import { LinearGaugeStylingDefaults } from './LinearGaugeStylingDefaults';
import { LinearScaleOrientation_$type } from './LinearScaleOrientation';
import { LinearGraphNeedleShape_$type } from './LinearGraphNeedleShape';
import { IgcFormatLinearGraphLabelEventArgs } from './igc-format-linear-graph-label-event-args';
import { IgcAlignLinearGraphLabelEventArgs } from './igc-align-linear-graph-label-event-args';
import { IgcLinearGraphRangeComponent } from './igc-linear-graph-range-component';
import { IgcLinearGraphRangeCollection } from './igc-linear-graph-range-collection';
import { RegisterElementHelper } from "igniteui-webcomponents-core";
import { IgcDoubleValueChangedEventArgs } from "igniteui-webcomponents-core";
import { IgcHTMLElement } from "igniteui-webcomponents-core";
var IgcLinearGaugeComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(IgcLinearGaugeComponent, _super);
    function IgcLinearGaugeComponent() {
        var _this = _super.call(this) || this;
        _this.contentRanges = [];
        /**
        * The ranges actually present in the chart. Do not directly modify this array.
        * This array's contents can be modified by causing Angular to reproject the child content.
        * Or adding and removing ranges from the manual ranges collection on the ranges property.
        */
        _this.actualRanges = [];
        _this._ranges = null;
        _this._rangesAdapter = null;
        _this._disconnected = false;
        _this._tooltipTemplate = null;
        _this._tooltipContent = null;
        _this._defaultTooltips = null;
        _this._settingAttributes = false;
        _this._attached = false;
        _this._queuedSetAttributes = [];
        _this._updatingFromAttribute = false;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._formatLabel = null;
        _this._formatLabel_wrapped = null;
        _this._alignLabel = null;
        _this._alignLabel_wrapped = null;
        _this._valueChanged = null;
        _this._valueChanged_wrapped = null;
        _this._actualMinimumValueChange = null;
        _this._actualMinimumValueChange_wrapped = null;
        _this._actualMaximumValueChange = null;
        _this._actualMaximumValueChange_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._renderer = new WebComponentRenderer(_this, document, true, LinearGaugeStylingDefaults);
        _this._implementation = _this.createImplementation();
        _this._container = _this._renderer.createElement("div");
        _this._renderer.updateRoot(_this._container);
        //this._renderer.rootWrapper.append(this._container);
        //this._renderer.rootWrapper.append(this._container);
        _this._container.setStyleProperty("display", "block");
        _this._container.setStyleProperty("width", "100%");
        _this._container.setStyleProperty("height", "100%");
        var root;
        root = _this._container;
        if (_this._container.getNativeElement() != null) {
            root = _this._container.getNativeElement();
        }
        _this._wrapper = _this._renderer;
        var gauge = _this.i;
        _this._gauge = gauge;
        _this._rangesAdapter = new CollectionAdapter(_this.contentRanges, _this.i.ranges, _this.actualRanges, function (c) { return c.i; }, function (i) {
            i._provideRenderer(_this._wrapper);
            if (_this.parentElement) {
                i._styling(_this, _this, _this);
            }
        }, function (i) { i._provideRenderer(null); });
        gauge.provideContainer(_this._renderer);
        var mut = new MutationObserver(function (list) {
            var e_1, _b;
            try {
                for (var list_1 = tslib_1.__values(list), list_1_1 = list_1.next(); !list_1_1.done; list_1_1 = list_1.next()) {
                    var mutation = list_1_1.value;
                    if (mutation.type == 'childList') {
                        _this.updateContentRanges();
                    }
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (list_1_1 && !list_1_1.done && (_b = list_1.return))
                        _b.call(list_1);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
        });
        mut.observe(_this, {
            childList: true
        });
        _this._renderer.addSizeWatcher(function () {
            _this._gauge.containerResized();
        });
        return _this;
    }
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "height", {
        get: function () {
            return this._height;
        },
        set: function (value) {
            this._height = value;
            this.style.height = value;
            this.containerResized();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "width", {
        get: function () {
            return this._width;
        },
        set: function (value) {
            this._width = value;
            this.style.width = value;
            this.containerResized();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "ranges", {
        /**
         * A collection or manually added axes for the chart.
        */
        get: function () {
            var _this = this;
            if (this._ranges === null) {
                var coll = new IgcLinearGraphRangeCollection();
                var inner = coll._innerColl;
                inner.addListener(function (sender, e) {
                    switch (e.action) {
                        case NotifyCollectionChangedAction.Add:
                            _this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Remove:
                            _this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            break;
                        case NotifyCollectionChangedAction.Replace:
                            _this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            _this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Reset:
                            _this._rangesAdapter.clearManualItems();
                            break;
                    }
                });
                this._ranges = coll;
            }
            return this._ranges;
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent.prototype.destroy = function () {
        this._wrapper.rootWrapper.unlistenAll();
        this._gauge.destroy();
        this._wrapper.destroy();
    };
    IgcLinearGaugeComponent.prototype.updateContentRanges = function () {
        this.contentRanges.length = 0;
        for (var i = 0; i < this.children.length; i++) {
            if (this.children[i] instanceof IgcLinearGraphRangeComponent) {
                this.contentRanges.push(this.children[i]);
            }
        }
        if (this._rangesAdapter != null) {
            this._rangesAdapter.notifyContentChanged();
        }
    };
    IgcLinearGaugeComponent.prototype.createImplementation = function () {
        return new XamLinearGauge();
    };
    IgcLinearGaugeComponent.prototype.disconnectedCallback = function () {
        this._disconnected = true;
        if (this.i) {
            this.i.onDetachedFromUI();
        }
    };
    IgcLinearGaugeComponent.prototype.connectedCallback = function () {
        if (this._disconnected) {
            this._disconnected = false;
            if (this.i) {
                this.i.onAttachedToUI();
            }
            return;
        }
        this.classList.add("ig-linear-gauge");
        this.classList.add("igc-linear-gauge");
        this.appendChild(this._renderer.rootWrapper.getNativeElement());
        this._attached = true;
        this.style.display = "block";
        this.style.height = this._height;
        this.style.width = this._width;
        this._flushQueuedAttributes();
        // supports themes or custom properties set in CSS 
        this._styling(this, this);
        for (var i = 0; i < this.actualRanges.length; i++) {
            var s = this.actualRanges[i];
            s._styling(this, this, this);
        }
        this.updateContentRanges();
        this.afterContentInit();
    };
    IgcLinearGaugeComponent.prototype.afterContentInit = function () {
        var _this = this;
        if (TypeRegistrar.isRegistered("IgcSimpleDefaultTooltipComponent")) {
            var cr_1 = TypeRegistrar.create("IgcSimpleDefaultTooltipComponent");
            this._defaultTooltips = cr_1;
            cr_1.onContentReady = function () {
                _this._onDefaultTooltipsReady(cr_1);
            };
            this._ensureTooltipCreated();
        }
        this.i.containerResized();
    };
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "tooltipTemplate", {
        get: function () {
            return this._tooltipTemplate;
        },
        set: function (value) {
            this._tooltipTemplate = value;
            if (value == null) {
                if (this._tooltipContent !== null) {
                    if (this._tooltipContent.destroy) {
                        this._tooltipContent.destroy();
                    }
                    this._tooltipContent = null;
                }
            }
            if (this._tooltipContent != null) {
                this._tooltipContent.template = this._tooltipTemplate;
            }
            else {
                this._ensureTooltipCreated();
            }
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent.prototype._ensureTooltipCreated = function () {
        if (this.i.toolTip == null) {
            var tooltip = this.createTooltip();
            if (tooltip == null) {
                return;
            }
            this._tooltipContent = tooltip;
            tooltip.template = this._tooltipTemplate;
            this.i.toolTip = this.createWrapper(tooltip);
        }
    };
    IgcLinearGaugeComponent.prototype.createWrapper = function (ele) {
        var wrapper = new WebComponentWrapper(ele, this._renderer);
        wrapper.updateToolTip = ele.updateToolTip;
        wrapper.hideToolTip = ele.hideToolTip;
        return wrapper;
    };
    IgcLinearGaugeComponent.prototype._ensureDefaultTooltip = function () {
        if (this._defaultTooltips == null) {
            return;
        }
        this._defaultTooltips["ensureDefaultTooltip"](this);
    };
    IgcLinearGaugeComponent.prototype._onDefaultTooltipsReady = function (cr) {
        this._ensureDefaultTooltip();
    };
    IgcLinearGaugeComponent.prototype.createTooltip = function () {
        if (!TypeRegistrar.isRegistered("IgcTooltipContainerComponent")) {
            return null;
        }
        var cr = TypeRegistrar.create("IgcTooltipContainerComponent");
        var ele = cr;
        var self = this;
        ele.updateToolTip = function (c) {
            if (c.externalObject) {
                c = c.externalObject;
            }
            else {
                var ext = new IgcDataContext();
                ext._implementation = c;
                c = ext;
            }
            if (ele.parentElement != self._container.getNativeElement()) {
                if (ele.parentElement != null) {
                    ele.parentElement.removeChild(ele);
                }
                self._container.append(this._renderer.createWrapper(ele));
            }
            cr.context = c;
            ele.style.display = "block";
            return true;
        };
        ele.hideToolTip = function () {
            ele.style.display = "none";
        };
        ele.style.display = "none";
        return cr;
    };
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igc" + name + "Component";
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgcLinearGaugeComponent.prototype._enqueueSetAttribute = function (attrName, attrValue) {
        this._queuedSetAttributes.push({ attrName: attrName, attrValue: attrValue });
    };
    IgcLinearGaugeComponent.prototype._flushQueuedAttributes = function () {
        this._settingAttributes = true;
        for (var i = 0; i < this._queuedSetAttributes.length; i++) {
            this.setAttribute(this._queuedSetAttributes[i].attrName, this._queuedSetAttributes[i].attrValue);
        }
        this._settingAttributes = false;
        this._queuedSetAttributes.length = 0;
    };
    IgcLinearGaugeComponent.prototype._a = function (attrName, attrValue) {
        if (this._updatingFromAttribute) {
            return;
        }
        if (attrValue) {
            attrValue = attrValue.toString();
        }
        this._settingAttributes = true;
        attrName = toSpinal(attrName);
        if (this._attached) {
            this.setAttribute(attrName, attrValue);
        }
        else {
            this._enqueueSetAttribute(attrName, attrValue);
        }
        this._settingAttributes = false;
    };
    Object.defineProperty(IgcLinearGaugeComponent, "observedAttributes", {
        get: function () {
            if (IgcLinearGaugeComponent._observedAttributesIgcLinearGaugeComponent == null) {
                var names = getAllPropertyNames(IgcLinearGaugeComponent);
                for (var i = 0; i < names.length; i++) {
                    names[i] = toSpinal(names[i]);
                }
                IgcLinearGaugeComponent._observedAttributesIgcLinearGaugeComponent = names;
            }
            return IgcLinearGaugeComponent._observedAttributesIgcLinearGaugeComponent;
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent.prototype.attributeChangedCallback = function (name, oldValue, newValue) {
        if (this._settingAttributes) {
            return;
        }
        var setName = fromSpinal(name);
        this._updatingFromAttribute = true;
        this[setName] = newValue;
        this._updatingFromAttribute = false;
    };
    IgcLinearGaugeComponent.register = function () {
        if (!IgcLinearGaugeComponent._isElementRegistered) {
            IgcLinearGaugeComponent._isElementRegistered = true;
            RegisterElementHelper.registerElement(IgcLinearGaugeComponent.htmlTagName, IgcLinearGaugeComponent);
        }
    };
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "animating", {
        /**
     * Gets a value indicating whether the bullet graph is currently animating.
    */
        get: function () {
            return this.i.a4;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "transitionProgress", {
        /**
         * Gets the transition progress of the animation when the control is animating.
        */
        get: function () {
            return this.i.cz;
        },
        set: function (v) {
            this.i.cz = +v;
            this._a("transitionProgress", this.i.cz);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "orientation", {
        /**
         * Gets or sets the orientation of the scale.
        */
        get: function () {
            return this.i.as;
        },
        set: function (v) {
            this.i.as = ensureEnum(LinearScaleOrientation_$type, v);
            this._a("orientation", enumToString(LinearScaleOrientation_$type, this.i.as));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "rangeBrushes", {
        /**
         * Gets or sets a collection of brushes to be used as the palette for linear gauge ranges.
        */
        get: function () {
            return fromBrushCollection(this.i.ah);
        },
        set: function (v) {
            this.i.ah = toBrushCollection(v);
            this._a("rangeBrushes", brushCollectionToString(this.i.ah));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "rangeOutlines", {
        /**
         * Gets or sets a collection of brushes to be used as the palette for linear gauge outlines.
        */
        get: function () {
            return fromBrushCollection(this.i.ai);
        },
        set: function (v) {
            this.i.ai = toBrushCollection(v);
            this._a("rangeOutlines", brushCollectionToString(this.i.ai));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minimumValue", {
        /**
         * Gets or sets the minimum value of the scale.
        */
        get: function () {
            return this.i.b4;
        },
        set: function (v) {
            this.i.b4 = +v;
            this._a("minimumValue", this.i.b4);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "actualMinimumValue", {
        /**
         * Gets the resolved minimum value of the scale.
        */
        get: function () {
            return this.i.bs;
        },
        set: function (v) {
            this.i.bs = +v;
            this._a("actualMinimumValue", this.i.bs);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "maximumValue", {
        /**
         * Gets or sets the maximum value of the scale.
        */
        get: function () {
            return this.i.b3;
        },
        set: function (v) {
            this.i.b3 = +v;
            this._a("maximumValue", this.i.b3);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "actualMaximumValue", {
        /**
         * Gets the resolved maximum value of the scale.
        */
        get: function () {
            return this.i.br;
        },
        set: function (v) {
            this.i.br = +v;
            this._a("actualMaximumValue", this.i.br);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "value", {
        /**
         * Gets or sets the value at which the needle is positioned.
        */
        get: function () {
            return this.i.c0;
        },
        set: function (v) {
            this.i.c0 = +v;
            this._a("value", this.i.c0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleShape", {
        /**
         * Gets or sets the shape to use when rendering the needle from a number of options.
        */
        get: function () {
            return this.i.ap;
        },
        set: function (v) {
            this.i.ap = ensureEnum(LinearGraphNeedleShape_$type, v);
            this._a("needleShape", enumToString(LinearGraphNeedleShape_$type, this.i.ap));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleName", {
        /**
         * Gets or sets the name used for needle.
        */
        get: function () {
            return this.i.d5;
        },
        set: function (v) {
            this.i.d5 = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "rangeInnerExtent", {
        /**
         * Gets or sets the position at which to start rendering the ranges, measured from the front/bottom of the control as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.ck;
        },
        set: function (v) {
            this.i.ck = +v;
            this._a("rangeInnerExtent", this.i.ck);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleInnerExtent", {
        /**
         * Gets or sets the position at which to start rendering the scale, measured from the bottom/front (when orientation is horizontal/vertical) of the control as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.co;
        },
        set: function (v) {
            this.i.co = +v;
            this._a("scaleInnerExtent", this.i.co);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "rangeOuterExtent", {
        /**
         * Gets or sets the position at which to stop rendering the range as a value from 0 to 1 measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cl;
        },
        set: function (v) {
            this.i.cl = +v;
            this._a("rangeOuterExtent", this.i.cl);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleOuterExtent", {
        /**
         * Gets or sets the position at which to stop rendering the scale as a value from 0 to 1 measured from the bottom/front (when orientation is horizontal/vertical) of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cq;
        },
        set: function (v) {
            this.i.cq = +v;
            this._a("scaleOuterExtent", this.i.cq);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleInnerExtent", {
        /**
         * Gets or sets the position at which to start rendering the needle geometry, measured from the front/bottom of the linear gauge as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cb;
        },
        set: function (v) {
            this.i.cb = +v;
            this._a("needleInnerExtent", this.i.cb);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleOuterExtent", {
        /**
         * Gets or sets the position at which to stop rendering the needle geometry as a value from 0 to 1 measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cf;
        },
        set: function (v) {
            this.i.cf = +v;
            this._a("needleOuterExtent", this.i.cf);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleInnerBaseWidth", {
        /**
         * Gets or sets the width of the needle's inner base.
        */
        get: function () {
            return this.i.ca;
        },
        set: function (v) {
            this.i.ca = +v;
            this._a("needleInnerBaseWidth", this.i.ca);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleOuterBaseWidth", {
        /**
         * Gets or sets the width of the needle's outer base.
        */
        get: function () {
            return this.i.ce;
        },
        set: function (v) {
            this.i.ce = +v;
            this._a("needleOuterBaseWidth", this.i.ce);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleInnerPointWidth", {
        /**
         * Gets or sets the width of the needle's inner point.
        */
        get: function () {
            return this.i.cd;
        },
        set: function (v) {
            this.i.cd = +v;
            this._a("needleInnerPointWidth", this.i.cd);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleOuterPointWidth", {
        /**
         * Gets or sets the width of the needle's outer point.
        */
        get: function () {
            return this.i.ch;
        },
        set: function (v) {
            this.i.ch = +v;
            this._a("needleOuterPointWidth", this.i.ch);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleInnerPointExtent", {
        /**
         * Gets or sets the extent of the needle's inner point.
        */
        get: function () {
            return this.i.cc;
        },
        set: function (v) {
            this.i.cc = +v;
            this._a("needleInnerPointExtent", this.i.cc);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleOuterPointExtent", {
        /**
         * Gets or sets the extent of the needle's outer point.
        */
        get: function () {
            return this.i.cg;
        },
        set: function (v) {
            this.i.cg = +v;
            this._a("needleOuterPointExtent", this.i.cg);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "interval", {
        /**
         * Gets or sets the interval to use for the scale.
        */
        get: function () {
            return this.i.by;
        },
        set: function (v) {
            this.i.by = +v;
            this._a("interval", this.i.by);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "ticksPostInitial", {
        /**
         * A value to start adding tickmarks, added to the scale's MinimumValue.
        */
        get: function () {
            return this.i.cv;
        },
        set: function (v) {
            this.i.cv = +v;
            this._a("ticksPostInitial", this.i.cv);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "ticksPreTerminal", {
        /**
         * A value to stop adding tickmarks, subtracted from the scale's MaximumValue.
        */
        get: function () {
            return this.i.cw;
        },
        set: function (v) {
            this.i.cw = +v;
            this._a("ticksPreTerminal", this.i.cw);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelInterval", {
        /**
         * Gets or sets the interval to use for rendering labels. This defaults to be the same interval as the tickmarks on the scale.
        */
        get: function () {
            return this.i.b0;
        },
        set: function (v) {
            this.i.b0 = +v;
            this._a("labelInterval", this.i.b0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelExtent", {
        /**
         * Gets or sets the position at which to put the labels as a value from 0 to 1, measured from the bottom of the scale.
         * Values further from zero than 1 can be used to hide the labels of the linear gauge.
        */
        get: function () {
            return this.i.bz;
        },
        set: function (v) {
            this.i.bz = +v;
            this._a("labelExtent", this.i.bz);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelsPostInitial", {
        /**
         * A value to start adding labels, added to the scale's MinimumValue.
        */
        get: function () {
            return this.i.b1;
        },
        set: function (v) {
            this.i.b1 = +v;
            this._a("labelsPostInitial", this.i.b1);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelsPreTerminal", {
        /**
         * A value to stop adding labels, subtracted from the scale's MaximumValue.
        */
        get: function () {
            return this.i.b2;
        },
        set: function (v) {
            this.i.b2 = +v;
            this._a("labelsPreTerminal", this.i.b2);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minorTickCount", {
        /**
         * Gets or sets the number of minor tickmarks to place between major tickmarks.
        */
        get: function () {
            return this.i.b5;
        },
        set: function (v) {
            this.i.b5 = +v;
            this._a("minorTickCount", this.i.b5);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "tickStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the major tickmarks as a value from 0 to 1, measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cx;
        },
        set: function (v) {
            this.i.cx = +v;
            this._a("tickStartExtent", this.i.cx);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "tickEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the major tickmarks as a value from 0 to 1, measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cu;
        },
        set: function (v) {
            this.i.cu = +v;
            this._a("tickEndExtent", this.i.cu);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "tickStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use when rendering ticks.
        */
        get: function () {
            return this.i.cy;
        },
        set: function (v) {
            this.i.cy = +v;
            this._a("tickStrokeThickness", this.i.cy);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "tickBrush", {
        /**
         * Gets or sets the brush to use for the major tickmarks.
        */
        get: function () {
            return brushToString(this.i.h1);
        },
        set: function (v) {
            this.i.h1 = stringToBrush(v);
            this._a("tickBrush", brushToString(this.i.h1));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "fontBrush", {
        /**
         * Gets or sets the brush to use for the label font.
        */
        get: function () {
            return brushToString(this.i.hu);
        },
        set: function (v) {
            this.i.hu = stringToBrush(v);
            this._a("fontBrush", brushToString(this.i.hu));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleBreadth", {
        /**
         * Gets or sets the needle breadth.
        */
        get: function () {
            return this.i.b9;
        },
        set: function (v) {
            this.i.b9 = +v;
            this._a("needleBreadth", this.i.b9);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleBrush", {
        /**
         * Gets or sets the brush to use for needle element.
        */
        get: function () {
            return brushToString(this.i.hx);
        },
        set: function (v) {
            this.i.hx = stringToBrush(v);
            this._a("needleBrush", brushToString(this.i.hx));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleOutline", {
        /**
         * Gets or sets the brush to use for the outline of needle element.
        */
        get: function () {
            return brushToString(this.i.hy);
        },
        set: function (v) {
            this.i.hy = stringToBrush(v);
            this._a("needleOutline", brushToString(this.i.hy));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "needleStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use when rendering single actual value element.
        */
        get: function () {
            return this.i.ci;
        },
        set: function (v) {
            this.i.ci = +v;
            this._a("needleStrokeThickness", this.i.ci);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minorTickStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the minor tickmarks as a value from 0 to 1, measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.b7;
        },
        set: function (v) {
            this.i.b7 = +v;
            this._a("minorTickStartExtent", this.i.b7);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minorTickEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the minor tickmarks as a value from 0 to 1, measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.b6;
        },
        set: function (v) {
            this.i.b6 = +v;
            this._a("minorTickEndExtent", this.i.b6);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minorTickStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use when rendering minor ticks.
        */
        get: function () {
            return this.i.b8;
        },
        set: function (v) {
            this.i.b8 = +v;
            this._a("minorTickStrokeThickness", this.i.b8);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "minorTickBrush", {
        /**
         * Gets or sets the brush to use for the minor tickmarks.
        */
        get: function () {
            return brushToString(this.i.hw);
        },
        set: function (v) {
            this.i.hw = stringToBrush(v);
            this._a("minorTickBrush", brushToString(this.i.hw));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "isScaleInverted", {
        /**
         * Gets or sets a value indicating whether the scale is inverted.
         * When the scale is inverted the direction in which the scale values increase is right to left.
        */
        get: function () {
            return this.i.a8;
        },
        set: function (v) {
            this.i.a8 = ensureBool(v);
            this._a("isScaleInverted", this.i.a8);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "backingBrush", {
        /**
         * Gets or sets the brush to use to fill the backing of the linear gauge.
        */
        get: function () {
            return brushToString(this.i.hs);
        },
        set: function (v) {
            this.i.hs = stringToBrush(v);
            this._a("backingBrush", brushToString(this.i.hs));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "backingOutline", {
        /**
         * Gets or sets the brush to use for the outline of the backing.
        */
        get: function () {
            return brushToString(this.i.ht);
        },
        set: function (v) {
            this.i.ht = stringToBrush(v);
            this._a("backingOutline", brushToString(this.i.ht));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "backingStrokeThickness", {
        /**
         * Gets or sets the stroke thickness of the backing outline.
        */
        get: function () {
            return this.i.bw;
        },
        set: function (v) {
            this.i.bw = +v;
            this._a("backingStrokeThickness", this.i.bw);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "backingInnerExtent", {
        /**
         * Gets or sets the inner extent of the linear gauge backing.
        */
        get: function () {
            return this.i.bu;
        },
        set: function (v) {
            this.i.bu = +v;
            this._a("backingInnerExtent", this.i.bu);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "backingOuterExtent", {
        /**
         * Gets or sets the outer extent of the linear gauge backing.
        */
        get: function () {
            return this.i.bv;
        },
        set: function (v) {
            this.i.bv = +v;
            this._a("backingOuterExtent", this.i.bv);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the scale, measured from the front/bottom of the linear gauge as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cs;
        },
        set: function (v) {
            this.i.cs = +v;
            this._a("scaleStartExtent", this.i.cs);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the scale as a value from 0 to 1, measured from the front/bottom of the linear gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal size of the linear gauge.
        */
        get: function () {
            return this.i.cn;
        },
        set: function (v) {
            this.i.cn = +v;
            this._a("scaleEndExtent", this.i.cn);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleBrush", {
        /**
         * Gets or sets the brush to use to fill the scale of the linear gauge.
        */
        get: function () {
            return brushToString(this.i.hz);
        },
        set: function (v) {
            this.i.hz = stringToBrush(v);
            this._a("scaleBrush", brushToString(this.i.hz));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleOutline", {
        /**
         * Gets or sets the brush to use for the outline of the scale.
        */
        get: function () {
            return brushToString(this.i.h0);
        },
        set: function (v) {
            this.i.h0 = stringToBrush(v);
            this._a("scaleOutline", brushToString(this.i.h0));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "scaleStrokeThickness", {
        /**
         * Gets or sets the stroke thickness of the scale outline.
        */
        get: function () {
            return this.i.ct;
        },
        set: function (v) {
            this.i.ct = +v;
            this._a("scaleStrokeThickness", this.i.ct);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "isNeedleDraggingEnabled", {
        /**
         * Gets or sets whether needle dragging is enabled or not.
        */
        get: function () {
            return this.i.a7;
        },
        set: function (v) {
            this.i.a7 = ensureBool(v);
            this._a("isNeedleDraggingEnabled", this.i.a7);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelFormat", {
        /**
         * Gets or sets the label composite format used when creating label values.
        */
        get: function () {
            return this.i.dk;
        },
        set: function (v) {
            this.i.dk = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "labelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the LabelFormat string.
        */
        get: function () {
            return this.i.ag;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ag = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "transitionDuration", {
        /**
         * Gets or sets the number of milliseconds over which changes to the linear gauge should be animated.
        */
        get: function () {
            return this.i.c4;
        },
        set: function (v) {
            this.i.c4 = +v;
            this._a("transitionDuration", this.i.c4);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "showToolTipTimeout", {
        /**
         * Gets or sets the time in milliseconds that tooltip appearance is delayed with.
        */
        get: function () {
            return this.i.c3;
        },
        set: function (v) {
            this.i.c3 = +v;
            this._a("showToolTipTimeout", this.i.c3);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "showToolTip", {
        /**
         * Gets or sets a value indicating whether tooltips are enabled.
        */
        get: function () {
            return this.i.be;
        },
        set: function (v) {
            this.i.be = ensureBool(v);
            this._a("showToolTip", this.i.be);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "font", {
        /**
         * Gets or sets the font.
        */
        get: function () {
            return this.i.dd;
        },
        set: function (v) {
            this.i.dd = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "pixelScalingRatio", {
        /**
         * Gets or sets the scaling value used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get: function () {
            return this.i.cj;
        },
        set: function (v) {
            this.i.cj = +v;
            this._a("pixelScalingRatio", this.i.cj);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "actualPixelScalingRatio", {
        /**
         * Gets the actual pixel scaling ratio used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get: function () {
            return this.i.bt;
        },
        set: function (v) {
            this.i.bt = +v;
            this._a("actualPixelScalingRatio", this.i.bt);
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.ranges != null && this.ranges.findByName && this.ranges.findByName(name)) {
            return this.ranges.findByName(name);
        }
        if (this.labelFormatSpecifiers != null && arrayFindByName(this.labelFormatSpecifiers, name)) {
            return arrayFindByName(this.labelFormatSpecifiers, name);
        }
        return null;
    };
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgcLinearGaugeComponent.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("LinearGaugeComponent");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    /**
 * Gets the value for the main scale of the gauge for a given point within the bounds of the gauge.

* @param point  * The point for which to retrieve the associated value.
*/
    IgcLinearGaugeComponent.prototype.getValueForPoint = function (point) {
        var iv = this.i.bx(toPoint(point));
        return (iv);
    };
    IgcLinearGaugeComponent.prototype.needleContainsPoint = function (point, isFinger) {
        var iv = this.i.needleContainsPoint(toPoint(point), isFinger);
        return (iv);
    };
    /**
     * Returns visuals as a serialized string.
    
    */
    IgcLinearGaugeComponent.prototype.exportSerializedVisualData = function () {
        var iv = this.i.dc();
        return (iv);
    };
    /**
     * Use to force the linear gauge to finish any deferred work before printing or evaluating its visual.
     * This should only be called if the visual of the linear gauge needs to be synchronously saved or evaluated.
     * Calling this method too often will hinder the performance of the linear gauge.
    
    */
    IgcLinearGaugeComponent.prototype.flush = function () {
        this.i.e8();
    };
    /**
     * Called by the UI framework to provide a UI container for rendering this control.
    
    * @param container  * The UI container element.
    */
    IgcLinearGaugeComponent.prototype.provideContainer = function (container) {
        this.i.provideContainer(container);
    };
    IgcLinearGaugeComponent.prototype.containerResized = function () {
        this.i.containerResized();
    };
    IgcLinearGaugeComponent.prototype.styleUpdated = function () {
        this.i.f2();
    };
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "formatLabel", {
        /**
         * Event which is raised when a label of the the gauge is formatted.
         * Function takes first argument evt and second argument ui.
         * Use ui.owner to obtain reference to the gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of the gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of the gauge scale.
         * Use ui.value to obtain the value on the the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
        */
        get: function () {
            return this._formatLabel;
        },
        set: function (ev) {
            var _this = this;
            if (this._formatLabel_wrapped !== null) {
                this.i.formatLabel = delegateRemove(this.i.formatLabel, this._formatLabel_wrapped);
                this._formatLabel_wrapped = null;
                this._formatLabel = null;
            }
            this._formatLabel = ev;
            this._formatLabel_wrapped = function (o, e) {
                var outerArgs = new IgcFormatLinearGraphLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeFormatLabel) {
                    _this.beforeFormatLabel(_this, outerArgs);
                }
                if (_this._formatLabel) {
                    _this._formatLabel(_this, outerArgs);
                }
            };
            this.i.formatLabel = delegateCombine(this.i.formatLabel, this._formatLabel_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "alignLabel", {
        /**
         * Event which is raised when a label of the linear gauge is aligned along the scale.
         * Function takes first argument evt and second argument ui.
         * Use ui.owner to obtain reference to the gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of the gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of the gauge scale.
         * Use ui.value to obtain the value on the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
         * Use ui.width to obtain the width of the label.
         * Use ui.height to obtain the height of the label.
         * Use ui.offsetX to obtain the X offset of the label on the gauge scale.
         * Use ui.offsetY to obtain the Y offset of the label on the gauge scale.
        */
        get: function () {
            return this._alignLabel;
        },
        set: function (ev) {
            var _this = this;
            if (this._alignLabel_wrapped !== null) {
                this.i.alignLabel = delegateRemove(this.i.alignLabel, this._alignLabel_wrapped);
                this._alignLabel_wrapped = null;
                this._alignLabel = null;
            }
            this._alignLabel = ev;
            this._alignLabel_wrapped = function (o, e) {
                var outerArgs = new IgcAlignLinearGraphLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeAlignLabel) {
                    _this.beforeAlignLabel(_this, outerArgs);
                }
                if (_this._alignLabel) {
                    _this._alignLabel(_this, outerArgs);
                }
            };
            this.i.alignLabel = delegateCombine(this.i.alignLabel, this._alignLabel_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "valueChanged", {
        /**
         * Occurs when the Value property changes.
        */
        get: function () {
            return this._valueChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._valueChanged_wrapped !== null) {
                this.i.valueChanged = delegateRemove(this.i.valueChanged, this._valueChanged_wrapped);
                this._valueChanged_wrapped = null;
                this._valueChanged = null;
            }
            this._valueChanged = ev;
            this._valueChanged_wrapped = function (o, e) {
                var outerArgs = new IgcDoubleValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeValueChanged) {
                    _this.beforeValueChanged(_this, outerArgs);
                }
                if (_this._valueChanged) {
                    _this._valueChanged(_this, outerArgs);
                }
            };
            this.i.valueChanged = delegateCombine(this.i.valueChanged, this._valueChanged_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "actualMinimumValueChange", {
        get: function () {
            return this._actualMinimumValueChange;
        },
        set: function (ev) {
            var _this = this;
            if (this._actualMinimumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
                this._actualMinimumValueChange_wrapped = null;
                this._actualMinimumValueChange = null;
            }
            this._actualMinimumValueChange = ev;
            this._actualMinimumValueChange_wrapped = function (o, e) {
                var ext = _this.actualMinimumValue;
                if (e.propertyName == 'Bs') {
                    if (_this.beforeActualMinimumValueChange) {
                        _this.beforeActualMinimumValueChange(_this, ext);
                    }
                    if (_this._actualMinimumValueChange) {
                        _this._actualMinimumValueChange(_this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcLinearGaugeComponent.prototype, "actualMaximumValueChange", {
        get: function () {
            return this._actualMaximumValueChange;
        },
        set: function (ev) {
            var _this = this;
            if (this._actualMaximumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
                this._actualMaximumValueChange_wrapped = null;
                this._actualMaximumValueChange = null;
            }
            this._actualMaximumValueChange = ev;
            this._actualMaximumValueChange_wrapped = function (o, e) {
                var ext = _this.actualMaximumValue;
                if (e.propertyName == 'Br') {
                    if (_this.beforeActualMaximumValueChange) {
                        _this.beforeActualMaximumValueChange(_this, ext);
                    }
                    if (_this._actualMaximumValueChange) {
                        _this._actualMaximumValueChange(_this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
        },
        enumerable: true,
        configurable: true
    });
    IgcLinearGaugeComponent._observedAttributesIgcLinearGaugeComponent = null;
    IgcLinearGaugeComponent.htmlTagName = "igc-linear-gauge";
    IgcLinearGaugeComponent._isElementRegistered = false;
    return IgcLinearGaugeComponent;
}(IgcHTMLElement));
export { IgcLinearGaugeComponent };
