// igniteui-webcomponents-gauges/esm5/lib/igc-radial-gauge-component.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { WebComponentRenderer } from "igniteui-webcomponents-core";
import { TypeRegistrar } from "igniteui-webcomponents-core";
import { XamRadialGauge } from './XamRadialGauge';
import { CollectionAdapter, brushToString, stringToBrush, toPoint, fromPoint, fromBrushCollection, toBrushCollection, ensureBool, ensureEnum, initializePropertiesFromCss, NamePatcher, toSpinal, fromSpinal, enumToString, brushCollectionToString, getAllPropertyNames, arrayFindByName } from "igniteui-webcomponents-core";
import { delegateCombine, delegateRemove } from "igniteui-webcomponents-core";
import { NotifyCollectionChangedAction } from "igniteui-webcomponents-core";
import { SweepDirection_$type } from "igniteui-webcomponents-core";
import { RadialGaugeNeedleShape_$type } from './RadialGaugeNeedleShape';
import { RadialGaugePivotShape_$type } from './RadialGaugePivotShape';
import { RadialGaugeScaleOversweepShape_$type } from './RadialGaugeScaleOversweepShape';
import { RadialGaugeBackingShape_$type } from './RadialGaugeBackingShape';
import { RadialGaugeDuplicateLabelOmissionStrategy_$type } from './RadialGaugeDuplicateLabelOmissionStrategy';
import { IgcFormatRadialGaugeLabelEventArgs } from './igc-format-radial-gauge-label-event-args';
import { IgcAlignRadialGaugeLabelEventArgs } from './igc-align-radial-gauge-label-event-args';
import { RadialGaugeStylingDefaults } from './RadialGaugeStylingDefaults';
import { IgcRadialGaugeRangeCollection } from './igc-radial-gauge-range-collection';
import { IgcRadialGaugeRangeComponent } from './igc-radial-gauge-range-component';
import { RegisterElementHelper } from "igniteui-webcomponents-core";
import { IgcDoubleValueChangedEventArgs } from "igniteui-webcomponents-core";
import { IgcHTMLElement } from "igniteui-webcomponents-core";
var IgcRadialGaugeComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(IgcRadialGaugeComponent, _super);
    function IgcRadialGaugeComponent() {
        var _this = _super.call(this) || this;
        _this.contentRanges = [];
        /**
        * The ranges actually present in the chart. Do not directly modify this array.
        * This array's contents can be modified by causing Angular to reproject the child content.
        * Or adding and removing ranges from the manual ranges collection on the ranges property.
        */
        _this.actualRanges = [];
        _this._ranges = null;
        _this._rangesAdapter = null;
        _this._disconnected = false;
        _this._settingAttributes = false;
        _this._attached = false;
        _this._queuedSetAttributes = [];
        _this._updatingFromAttribute = false;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._formatLabel = null;
        _this._formatLabel_wrapped = null;
        _this._alignLabel = null;
        _this._alignLabel_wrapped = null;
        _this._valueChanged = null;
        _this._valueChanged_wrapped = null;
        _this._actualMinimumValueChange = null;
        _this._actualMinimumValueChange_wrapped = null;
        _this._actualMaximumValueChange = null;
        _this._actualMaximumValueChange_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._renderer = new WebComponentRenderer(_this, document, true, RadialGaugeStylingDefaults);
        _this._implementation = _this.createImplementation();
        _this._container = _this._renderer.createElement("div");
        _this._renderer.updateRoot(_this._container);
        //this._renderer.rootWrapper.append(this._container);
        _this._container.setStyleProperty("display", "block");
        _this._container.setStyleProperty("width", "100%");
        _this._container.setStyleProperty("height", "100%");
        var root;
        root = _this._container;
        if (_this._container.getNativeElement() != null) {
            root = _this._container.getNativeElement();
        }
        _this._wrapper = _this._renderer;
        var gauge = _this.i;
        _this._gauge = gauge;
        _this._rangesAdapter = new CollectionAdapter(_this.contentRanges, _this.i.ranges, _this.actualRanges, function (c) { return c.i; }, function (i) {
            i._provideRenderer(_this._wrapper);
            if (_this.parentElement) {
                i._styling(_this, _this, _this);
            }
        }, function (i) { i._provideRenderer(null); });
        gauge.provideContainer(_this._renderer);
        var mut = new MutationObserver(function (list) {
            var e_1, _b;
            try {
                for (var list_1 = tslib_1.__values(list), list_1_1 = list_1.next(); !list_1_1.done; list_1_1 = list_1.next()) {
                    var mutation = list_1_1.value;
                    if (mutation.type == 'childList') {
                        _this.updateContentRanges();
                    }
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (list_1_1 && !list_1_1.done && (_b = list_1.return))
                        _b.call(list_1);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
        });
        mut.observe(_this, {
            childList: true
        });
        _this._renderer.addSizeWatcher(function () {
            _this._gauge.containerResized();
        });
        return _this;
    }
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "height", {
        get: function () {
            return this._height;
        },
        set: function (value) {
            this._height = value;
            this.style.height = value;
            this.containerResized();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "width", {
        get: function () {
            return this._width;
        },
        set: function (value) {
            this._width = value;
            this.style.width = value;
            this.containerResized();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "ranges", {
        /**
         * A collection or manually added axes for the chart.
        */
        get: function () {
            var _this = this;
            if (this._ranges === null) {
                var coll = new IgcRadialGaugeRangeCollection();
                var inner = coll._innerColl;
                inner.addListener(function (sender, e) {
                    switch (e.action) {
                        case NotifyCollectionChangedAction.Add:
                            _this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Remove:
                            _this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            break;
                        case NotifyCollectionChangedAction.Replace:
                            _this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                            _this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                            break;
                        case NotifyCollectionChangedAction.Reset:
                            _this._rangesAdapter.clearManualItems();
                            break;
                    }
                });
                this._ranges = coll;
            }
            return this._ranges;
        },
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent.prototype.destroy = function () {
        this._gauge.destroy();
        this._wrapper.destroy();
    };
    IgcRadialGaugeComponent.prototype.updateContentRanges = function () {
        this.contentRanges.length = 0;
        for (var i = 0; i < this.children.length; i++) {
            if (this.children[i] instanceof IgcRadialGaugeRangeComponent) {
                this.contentRanges.push(this.children[i]);
            }
        }
        if (this._rangesAdapter != null) {
            this._rangesAdapter.notifyContentChanged();
        }
    };
    IgcRadialGaugeComponent.prototype.createImplementation = function () {
        return new XamRadialGauge();
    };
    IgcRadialGaugeComponent.prototype.disconnectedCallback = function () {
        this._disconnected = true;
        if (this.i) {
            this.i.onDetachedFromUI();
        }
    };
    IgcRadialGaugeComponent.prototype.connectedCallback = function () {
        if (this._disconnected) {
            this._disconnected = false;
            if (this.i) {
                this.i.onAttachedToUI();
            }
            return;
        }
        this.classList.add("ig-radial-gauge");
        this.classList.add("igc-radial-gauge");
        this.appendChild(this._renderer.rootWrapper.getNativeElement());
        this._attached = true;
        this.style.display = "block";
        this.style.height = this._height;
        this.style.width = this._width;
        this._flushQueuedAttributes();
        // supports themes or custom properties set in CSS 
        this._styling(this, this);
        for (var i = 0; i < this.actualRanges.length; i++) {
            var s = this.actualRanges[i];
            s._styling(this, this, this);
        }
        this.updateContentRanges();
        this.afterContentInit();
    };
    IgcRadialGaugeComponent.prototype.afterContentInit = function () {
        this.i.containerResized();
    };
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igc" + name + "Component";
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgcRadialGaugeComponent.prototype._enqueueSetAttribute = function (attrName, attrValue) {
        this._queuedSetAttributes.push({ attrName: attrName, attrValue: attrValue });
    };
    IgcRadialGaugeComponent.prototype._flushQueuedAttributes = function () {
        this._settingAttributes = true;
        for (var i = 0; i < this._queuedSetAttributes.length; i++) {
            this.setAttribute(this._queuedSetAttributes[i].attrName, this._queuedSetAttributes[i].attrValue);
        }
        this._settingAttributes = false;
        this._queuedSetAttributes.length = 0;
    };
    IgcRadialGaugeComponent.prototype._a = function (attrName, attrValue) {
        if (this._updatingFromAttribute) {
            return;
        }
        if (attrValue) {
            attrValue = attrValue.toString();
        }
        this._settingAttributes = true;
        attrName = toSpinal(attrName);
        if (this._attached) {
            this.setAttribute(attrName, attrValue);
        }
        else {
            this._enqueueSetAttribute(attrName, attrValue);
        }
        this._settingAttributes = false;
    };
    Object.defineProperty(IgcRadialGaugeComponent, "observedAttributes", {
        get: function () {
            if (IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent == null) {
                var names = getAllPropertyNames(IgcRadialGaugeComponent);
                for (var i = 0; i < names.length; i++) {
                    names[i] = toSpinal(names[i]);
                }
                IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent = names;
            }
            return IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent;
        },
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent.prototype.attributeChangedCallback = function (name, oldValue, newValue) {
        if (this._settingAttributes) {
            return;
        }
        var setName = fromSpinal(name);
        this._updatingFromAttribute = true;
        this[setName] = newValue;
        this._updatingFromAttribute = false;
    };
    IgcRadialGaugeComponent.register = function () {
        if (!IgcRadialGaugeComponent._isElementRegistered) {
            IgcRadialGaugeComponent._isElementRegistered = true;
            RegisterElementHelper.registerElement(IgcRadialGaugeComponent.htmlTagName, IgcRadialGaugeComponent);
        }
    };
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "rangeBrushes", {
        /**
     * Gets or sets a collection of brushes to be used as the palette for gauge ranges.
    */
        get: function () {
            return fromBrushCollection(this.i.ah);
        },
        set: function (v) {
            this.i.ah = toBrushCollection(v);
            this._a("rangeBrushes", brushCollectionToString(this.i.ah));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "rangeOutlines", {
        /**
         * Gets or sets a collection of brushes to be used as the palette for gauge outlines.
        */
        get: function () {
            return fromBrushCollection(this.i.ai);
        },
        set: function (v) {
            this.i.ai = toBrushCollection(v);
            this._a("rangeOutlines", brushCollectionToString(this.i.ai));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minimumValue", {
        /**
         * Gets or sets the minimum value of the scale.
        */
        get: function () {
            return this.i.cc;
        },
        set: function (v) {
            this.i.cc = +v;
            this._a("minimumValue", this.i.cc);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "actualMinimumValue", {
        /**
         * Gets the resolved minimum value of the scale.
        */
        get: function () {
            return this.i.bs;
        },
        set: function (v) {
            this.i.bs = +v;
            this._a("actualMinimumValue", this.i.bs);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "maximumValue", {
        /**
         * Gets or sets the maximum value of the scale.
        */
        get: function () {
            return this.i.ca;
        },
        set: function (v) {
            this.i.ca = +v;
            this._a("maximumValue", this.i.ca);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "actualMaximumValue", {
        /**
         * Gets the resolved maximum value of the scale.
        */
        get: function () {
            return this.i.br;
        },
        set: function (v) {
            this.i.br = +v;
            this._a("actualMaximumValue", this.i.br);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "interval", {
        /**
         * Gets or sets the interval to use for the scale.
        */
        get: function () {
            return this.i.b6;
        },
        set: function (v) {
            this.i.b6 = +v;
            this._a("interval", this.i.b6);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "centerX", {
        /**
         * Gets or sets the x position of the center of the gauge with the value ranging from 0 to 1.
        */
        get: function () {
            return this.i.b3;
        },
        set: function (v) {
            this.i.b3 = +v;
            this._a("centerX", this.i.b3);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "centerY", {
        /**
         * Gets or sets the y position of the center of the gauge with the value ranging from 0 to 1.
        */
        get: function () {
            return this.i.b4;
        },
        set: function (v) {
            this.i.b4 = +v;
            this._a("centerY", this.i.b4);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "value", {
        /**
         * Gets or sets the value at which to point the needle of the gauge.
        */
        get: function () {
            return this.i.dm;
        },
        set: function (v) {
            this.i.dm = +v;
            this._a("value", this.i.dm);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleStartAngle", {
        /**
         * Gets or sets the start angle for the scale in degrees.
        */
        get: function () {
            return this.i.da;
        },
        set: function (v) {
            this.i.da = +v;
            this._a("scaleStartAngle", this.i.da);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleEndAngle", {
        /**
         * Gets or sets the end angle for the scale in degrees.
        */
        get: function () {
            return this.i.c5;
        },
        set: function (v) {
            this.i.c5 = +v;
            this._a("scaleEndAngle", this.i.c5);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleSweepDirection", {
        /**
         * Gets or sets the direction in which the scale sweeps around the center from the start angle to end angle.
        */
        get: function () {
            return this.i.ii;
        },
        set: function (v) {
            this.i.ii = ensureEnum(SweepDirection_$type, v);
            this._a("scaleSweepDirection", enumToString(SweepDirection_$type, this.i.ii));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "transitionDuration", {
        /**
         * Gets or sets the number of milliseconds over which changes to the gauge should be animated.
        */
        get: function () {
            return this.i.dn;
        },
        set: function (v) {
            this.i.dn = +v;
            this._a("transitionDuration", this.i.dn);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "transitionEasingFunction", {
        /**
         * Gets or sets the easing function used to morph the current series.
        */
        get: function () {
            return this.i.a0;
        },
        set: function (v) {
            this.i.a0 = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleBrush", {
        /**
         * Gets or sets the brush to use when rendering the fill of the needle.
        */
        get: function () {
            return brushToString(this.i.ib);
        },
        set: function (v) {
            this.i.ib = stringToBrush(v);
            this._a("needleBrush", brushToString(this.i.ib));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleOutline", {
        /**
         * Gets or sets the brush to use when rendering the outline of the needle.
        */
        get: function () {
            return brushToString(this.i.ic);
        },
        set: function (v) {
            this.i.ic = stringToBrush(v);
            this._a("needleOutline", brushToString(this.i.ic));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleStartExtent", {
        /**
         * Gets or sets the extent (from -1 to 1) at which to start rendering the needle, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cv;
        },
        set: function (v) {
            this.i.cv = +v;
            this._a("needleStartExtent", this.i.cv);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleEndExtent", {
        /**
         * Gets or sets the extent (from -1 to 1) at which to end rendering the needle, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cn;
        },
        set: function (v) {
            this.i.cn = +v;
            this._a("needleEndExtent", this.i.cn);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleShape", {
        /**
         * Gets or sets the shape to use when rendering the needle from a number of options.
        */
        get: function () {
            return this.i.ar;
        },
        set: function (v) {
            this.i.ar = ensureEnum(RadialGaugeNeedleShape_$type, v);
            this._a("needleShape", enumToString(RadialGaugeNeedleShape_$type, this.i.ar));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleStartWidthRatio", {
        /**
         * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cw;
        },
        set: function (v) {
            this.i.cw = +v;
            this._a("needleStartWidthRatio", this.i.cw);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleEndWidthRatio", {
        /**
         * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.co;
        },
        set: function (v) {
            this.i.co = +v;
            this._a("needleEndWidthRatio", this.i.co);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleBaseFeatureWidthRatio", {
        /**
         * Gets or sets the width of the needle at its feature which is closest to the base (e.g. a bulb) with a value from 0 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cm;
        },
        set: function (v) {
            this.i.cm = +v;
            this._a("needleBaseFeatureWidthRatio", this.i.cm);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleBaseFeatureExtent", {
        /**
         * Gets or sets the extent of the feature which is closest to the base (e.g. a bulb) with a value from -1 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cl;
        },
        set: function (v) {
            this.i.cl = +v;
            this._a("needleBaseFeatureExtent", this.i.cl);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePointFeatureWidthRatio", {
        /**
         * Gets or sets the width of the needle at its feature which is closest to the point (e.g. the tapering point of a needle) with a value from 0 to 1. Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cu;
        },
        set: function (v) {
            this.i.cu = +v;
            this._a("needlePointFeatureWidthRatio", this.i.cu);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePointFeatureExtent", {
        /**
         * Gets or sets the extent of the feature which is closest to the point (e.g. the tapering point of a needle) with a value from -1 to 1. Note: Only some needle shapes respect this property.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.ct;
        },
        set: function (v) {
            this.i.ct = +v;
            this._a("needlePointFeatureExtent", this.i.ct);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotWidthRatio", {
        /**
         * Gets or sets the width of the cap of the needle with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cs;
        },
        set: function (v) {
            this.i.cs = +v;
            this._a("needlePivotWidthRatio", this.i.cs);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotInnerWidthRatio", {
        /**
         * Gets or sets the width of the inner cutout section of the needle cap with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle that has a cutout section.
        */
        get: function () {
            return this.i.cp;
        },
        set: function (v) {
            this.i.cp = +v;
            this._a("needlePivotInnerWidthRatio", this.i.cp);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotShape", {
        /**
         * Gets or sets the shape to use for the needle cap.
        */
        get: function () {
            return this.i.as;
        },
        set: function (v) {
            this.i.as = ensureEnum(RadialGaugePivotShape_$type, v);
            this._a("needlePivotShape", enumToString(RadialGaugePivotShape_$type, this.i.as));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the scale, measured from the center of the gauge as a value from 0 to 1.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.db;
        },
        set: function (v) {
            this.i.db = +v;
            this._a("scaleStartExtent", this.i.db);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotBrush", {
        /**
         * Gets or sets the brush to use for filling the needle cap. Note: this only applies to certain cap shapes.
        */
        get: function () {
            return brushToString(this.i.id);
        },
        set: function (v) {
            this.i.id = stringToBrush(v);
            this._a("needlePivotBrush", brushToString(this.i.id));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotOutline", {
        /**
         * Gets or sets the brush to use for the outlines of the needle cap.
        */
        get: function () {
            return brushToString(this.i.ie);
        },
        set: function (v) {
            this.i.ie = stringToBrush(v);
            this._a("needlePivotOutline", brushToString(this.i.ie));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needleStrokeThickness", {
        /**
         * Gets or sets the stroke thickness of the needle outline.
        */
        get: function () {
            return this.i.cx;
        },
        set: function (v) {
            this.i.cx = +v;
            this._a("needleStrokeThickness", this.i.cx);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "needlePivotStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use for the outline of the needle cap.
        */
        get: function () {
            return this.i.cq;
        },
        set: function (v) {
            this.i.cq = +v;
            this._a("needlePivotStrokeThickness", this.i.cq);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the scale as a value from 0 to 1 measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.c6;
        },
        set: function (v) {
            this.i.c6 = +v;
            this._a("scaleEndExtent", this.i.c6);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "labelExtent", {
        /**
         * Gets or sets the position at which to put the labels as a value from 0 to 1, measured form the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.b7;
        },
        set: function (v) {
            this.i.b7 = +v;
            this._a("labelExtent", this.i.b7);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "labelInterval", {
        /**
         * Gets or sets the interval to use for rendering labels. This defaults to be the same interval as the tickmarks on the scale.
        */
        get: function () {
            return this.i.b9;
        },
        set: function (v) {
            this.i.b9 = +v;
            this._a("labelInterval", this.i.b9);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "tickStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.dg;
        },
        set: function (v) {
            this.i.dg = +v;
            this._a("tickStartExtent", this.i.dg);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "tickEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.de;
        },
        set: function (v) {
            this.i.de = +v;
            this._a("tickEndExtent", this.i.de);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "tickStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use when rendering ticks.
        */
        get: function () {
            return this.i.di;
        },
        set: function (v) {
            this.i.di = +v;
            this._a("tickStrokeThickness", this.i.di);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "tickBrush", {
        /**
         * Gets or sets the brush to use for the major tickmarks.
        */
        get: function () {
            return brushToString(this.i.ih);
        },
        set: function (v) {
            this.i.ih = stringToBrush(v);
            this._a("tickBrush", brushToString(this.i.ih));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "fontBrush", {
        /**
         * Gets or sets the brush to use for the label font.
        */
        get: function () {
            return brushToString(this.i.h8);
        },
        set: function (v) {
            this.i.h8 = stringToBrush(v);
            this._a("fontBrush", brushToString(this.i.h8));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minorTickStartExtent", {
        /**
         * Gets or sets the position at which to start rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.ch;
        },
        set: function (v) {
            this.i.ch = +v;
            this._a("minorTickStartExtent", this.i.ch);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minorTickEndExtent", {
        /**
         * Gets or sets the position at which to stop rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
         * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
        */
        get: function () {
            return this.i.cf;
        },
        set: function (v) {
            this.i.cf = +v;
            this._a("minorTickEndExtent", this.i.cf);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minorTickStrokeThickness", {
        /**
         * Gets or sets the stroke thickness to use when rendering minor ticks.
        */
        get: function () {
            return this.i.cj;
        },
        set: function (v) {
            this.i.cj = +v;
            this._a("minorTickStrokeThickness", this.i.cj);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minorTickBrush", {
        /**
         * Gets or sets the brush to use for the minor tickmarks.
        */
        get: function () {
            return brushToString(this.i.ia);
        },
        set: function (v) {
            this.i.ia = stringToBrush(v);
            this._a("minorTickBrush", brushToString(this.i.ia));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "minorTickCount", {
        /**
         * Gets or sets the number of minor tickmarks to place between major tickmarks.
        */
        get: function () {
            return this.i.ce;
        },
        set: function (v) {
            this.i.ce = +v;
            this._a("minorTickCount", this.i.ce);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleBrush", {
        /**
         * Gets or sets the brush to use to fill the background of the scale.
        */
        get: function () {
            return brushToString(this.i.ig);
        },
        set: function (v) {
            this.i.ig = stringToBrush(v);
            this._a("scaleBrush", brushToString(this.i.ig));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingBrush", {
        /**
         * Gets or sets the brush to use to fill the backing of the gauge.
        */
        get: function () {
            return brushToString(this.i.h6);
        },
        set: function (v) {
            this.i.h6 = stringToBrush(v);
            this._a("backingBrush", brushToString(this.i.h6));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingOutline", {
        /**
         * Gets or sets the brush to use for the outline of the backing.
        */
        get: function () {
            return brushToString(this.i.h7);
        },
        set: function (v) {
            this.i.h7 = stringToBrush(v);
            this._a("backingOutline", brushToString(this.i.h7));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingStrokeThickness", {
        /**
         * Gets or sets the stroke thickness of the backing outline.
        */
        get: function () {
            return this.i.b1;
        },
        set: function (v) {
            this.i.b1 = +v;
            this._a("backingStrokeThickness", this.i.b1);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingOuterExtent", {
        /**
         * Gets or sets the outer extent of the gauge backing.
        */
        get: function () {
            return this.i.by;
        },
        set: function (v) {
            this.i.by = +v;
            this._a("backingOuterExtent", this.i.by);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingOversweep", {
        /**
         * Gets or sets the over sweep angle to apply to the backing if it is displaying fitted (in degrees). Must be greater or equal to 0.
        */
        get: function () {
            return this.i.b0;
        },
        set: function (v) {
            this.i.b0 = +v;
            this._a("backingOversweep", this.i.b0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleOversweep", {
        /**
         * Gets or sets the extra degrees of sweep to apply to the scale background. Must be greater or equal to 0.
        */
        get: function () {
            return this.i.c8;
        },
        set: function (v) {
            this.i.c8 = +v;
            this._a("scaleOversweep", this.i.c8);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "scaleOversweepShape", {
        /**
         * Gets or sets the oversweep shape to use for the excess fill area for the scale.
        */
        get: function () {
            return this.i.au;
        },
        set: function (v) {
            this.i.au = ensureEnum(RadialGaugeScaleOversweepShape_$type, v);
            this._a("scaleOversweepShape", enumToString(RadialGaugeScaleOversweepShape_$type, this.i.au));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingCornerRadius", {
        /**
         * Gets or sets the corner rounding radius to use for the fitted scale backings.
        */
        get: function () {
            return this.i.bu;
        },
        set: function (v) {
            this.i.bu = +v;
            this._a("backingCornerRadius", this.i.bu);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingInnerExtent", {
        /**
         * Gets or sets the inner extent of the gauge backing.
        */
        get: function () {
            return this.i.bw;
        },
        set: function (v) {
            this.i.bw = +v;
            this._a("backingInnerExtent", this.i.bw);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "backingShape", {
        /**
         * Gets or sets the type of shape to use for the backing of the gauge.
        */
        get: function () {
            return this.i.am;
        },
        set: function (v) {
            this.i.am = ensureEnum(RadialGaugeBackingShape_$type, v);
            this._a("backingShape", enumToString(RadialGaugeBackingShape_$type, this.i.am));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "radiusMultiplier", {
        /**
         * Gets or sets the multiplying factor to apply to the normal radius of the gauge.
         * The radius of the gauge is defined by the minimum of the width and height of the control divided by 2.0.
         * This introduces a multiplicative factor to that value.
        */
        get: function () {
            return this.i.c0;
        },
        set: function (v) {
            this.i.c0 = +v;
            this._a("radiusMultiplier", this.i.c0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "duplicateLabelOmissionStrategy", {
        /**
         * Gets or sets the strategy to use for omitting labels if the first and last label have the same value.
        */
        get: function () {
            return this.i.an;
        },
        set: function (v) {
            this.i.an = ensureEnum(RadialGaugeDuplicateLabelOmissionStrategy_$type, v);
            this._a("duplicateLabelOmissionStrategy", enumToString(RadialGaugeDuplicateLabelOmissionStrategy_$type, this.i.an));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "isNeedleDraggingEnabled", {
        /**
         * Gets or sets whether needle dragging is enabled or not.
        */
        get: function () {
            return this.i.ba;
        },
        set: function (v) {
            this.i.ba = ensureBool(v);
            this._a("isNeedleDraggingEnabled", this.i.ba);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "isNeedleDraggingConstrained", {
        /**
         * Gets or sets whether the needle is constrained within the minimum and maximum value range during dragging.
        */
        get: function () {
            return this.i.a9;
        },
        set: function (v) {
            this.i.a9 = ensureBool(v);
            this._a("isNeedleDraggingConstrained", this.i.a9);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "font", {
        /**
         * Gets or sets the font.
        */
        get: function () {
            return this.i.d2;
        },
        set: function (v) {
            this.i.d2 = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "labelFormat", {
        /**
         * Gets or sets the label composite format used when creating label values.
        */
        get: function () {
            return this.i.d9;
        },
        set: function (v) {
            this.i.d9 = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "labelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the LabelFormat string.
        */
        get: function () {
            return this.i.ag;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ag = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "animating", {
        /**
         * Gets a value indicating whether the radial gauge is currently animating.
        */
        get: function () {
            return this.i.a6;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "transitionProgress", {
        /**
         * Gets the transition progress of the animation when the control is animating.
        */
        get: function () {
            return this.i.dk;
        },
        set: function (v) {
            this.i.dk = +v;
            this._a("transitionProgress", this.i.dk);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "pixelScalingRatio", {
        /**
         * Gets or sets the scaling value used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get: function () {
            return this.i.cz;
        },
        set: function (v) {
            this.i.cz = +v;
            this._a("pixelScalingRatio", this.i.cz);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "actualPixelScalingRatio", {
        /**
         * Gets the actual pixel scaling ratio used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get: function () {
            return this.i.bt;
        },
        set: function (v) {
            this.i.bt = +v;
            this._a("actualPixelScalingRatio", this.i.bt);
        },
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.ranges != null && this.ranges.findByName && this.ranges.findByName(name)) {
            return this.ranges.findByName(name);
        }
        if (this.labelFormatSpecifiers != null && arrayFindByName(this.labelFormatSpecifiers, name)) {
            return arrayFindByName(this.labelFormatSpecifiers, name);
        }
        return null;
    };
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgcRadialGaugeComponent.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("RadialGaugeComponent");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    /**
 * Scales a value on the gauge's main scale to an angle around the center point of the gauge, in radians.

* @param value  * The value to scale.
*/
    IgcRadialGaugeComponent.prototype.scaleValue = function (value) {
        var iv = this.i.dd(value);
        return (iv);
    };
    /**
     * Unscales a value from an angle in radians to the represented value along the main scale of the gauge.
    
    * @param angle  * The angle in radians to unscale to a value.
    */
    IgcRadialGaugeComponent.prototype.unscaleValue = function (angle) {
        var iv = this.i.dl(angle);
        return (iv);
    };
    /**
     * Gets the value for the main scale of the gauge for a given point within the bounds of the gauge.
    
    * @param point  * The point for which to retrieve the associated value.
    */
    IgcRadialGaugeComponent.prototype.getValueForPoint = function (point) {
        var iv = this.i.b5(toPoint(point));
        return (iv);
    };
    IgcRadialGaugeComponent.prototype.getPointForValue = function (value, extent) {
        var iv = this.i.ik(value, extent);
        return fromPoint(iv);
    };
    IgcRadialGaugeComponent.prototype.styleUpdated = function () {
        this.i.gd();
    };
    IgcRadialGaugeComponent.prototype.needleContainsPoint = function (point, isFinger) {
        var iv = this.i.needleContainsPoint(toPoint(point), isFinger);
        return (iv);
    };
    /**
     * Called by the UI framework to provide a UI container for rendering this control.
    
    * @param container  * The UI container element.
    */
    IgcRadialGaugeComponent.prototype.provideContainer = function (container) {
        this.i.provideContainer(container);
    };
    IgcRadialGaugeComponent.prototype.containerResized = function () {
        this.i.containerResized();
    };
    /**
     * Use to force the radial gauge to finish any deferred work before printing or evaluating its visual.
     * This should only be called if the visual of the radial gauge needs to be synchronously saved or evaluated.
     * Calling this method too often will hinder the performance of the radial gauge.
    
    */
    IgcRadialGaugeComponent.prototype.flush = function () {
        this.i.fs();
    };
    /**
     * Returns visuals as a serialized string.
    
    */
    IgcRadialGaugeComponent.prototype.exportSerializedVisualData = function () {
        var iv = this.i.d1();
        return (iv);
    };
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "formatLabel", {
        /**
         * Event which is raised when a label of the gauge is formatted.
         * Function takes first argument null and second argument ui.
         * Use ui.owner to obtain reference to gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
         * Use ui.startAngle to obtain the starting angle of gauge scale.
         * Use ui.endAngle to obtain the ending angle of gauge scale.
         * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
         * Use ui.value to obtain the value on the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
        */
        get: function () {
            return this._formatLabel;
        },
        set: function (ev) {
            var _this = this;
            if (this._formatLabel_wrapped !== null) {
                this.i.formatLabel = delegateRemove(this.i.formatLabel, this._formatLabel_wrapped);
                this._formatLabel_wrapped = null;
                this._formatLabel = null;
            }
            this._formatLabel = ev;
            this._formatLabel_wrapped = function (o, e) {
                var outerArgs = new IgcFormatRadialGaugeLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeFormatLabel) {
                    _this.beforeFormatLabel(_this, outerArgs);
                }
                if (_this._formatLabel) {
                    _this._formatLabel(_this, outerArgs);
                }
            };
            this.i.formatLabel = delegateCombine(this.i.formatLabel, this._formatLabel_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "alignLabel", {
        /**
         * Event which is raised when a label of the gauge is aligned along the scale.
         * Function takes first argument null and second argument ui.
         * Use ui.owner to obtain reference to gauge widget.
         * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
         * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
         * Use ui.startAngle to obtain the starting angle of gauge scale.
         * Use ui.endAngle to obtain the ending angle of gauge scale.
         * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
         * Use ui.value to obtain the value on the gauge scale associated with the label.
         * Use ui.label to obtain the string value of the label.
         * Use ui.width to obtain the width of the label.
         * Use ui.height to obtain the height of the label.
         * Use ui.offsetX to obtain the X offset of the label on the gauge scale.
         * Use ui.offsetY to obtain the Y offset of the label on the gauge scale.
        */
        get: function () {
            return this._alignLabel;
        },
        set: function (ev) {
            var _this = this;
            if (this._alignLabel_wrapped !== null) {
                this.i.alignLabel = delegateRemove(this.i.alignLabel, this._alignLabel_wrapped);
                this._alignLabel_wrapped = null;
                this._alignLabel = null;
            }
            this._alignLabel = ev;
            this._alignLabel_wrapped = function (o, e) {
                var outerArgs = new IgcAlignRadialGaugeLabelEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeAlignLabel) {
                    _this.beforeAlignLabel(_this, outerArgs);
                }
                if (_this._alignLabel) {
                    _this._alignLabel(_this, outerArgs);
                }
            };
            this.i.alignLabel = delegateCombine(this.i.alignLabel, this._alignLabel_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "valueChanged", {
        /**
         * Occurs when the Value property changes.
        */
        get: function () {
            return this._valueChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._valueChanged_wrapped !== null) {
                this.i.valueChanged = delegateRemove(this.i.valueChanged, this._valueChanged_wrapped);
                this._valueChanged_wrapped = null;
                this._valueChanged = null;
            }
            this._valueChanged = ev;
            this._valueChanged_wrapped = function (o, e) {
                var outerArgs = new IgcDoubleValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeValueChanged) {
                    _this.beforeValueChanged(_this, outerArgs);
                }
                if (_this._valueChanged) {
                    _this._valueChanged(_this, outerArgs);
                }
            };
            this.i.valueChanged = delegateCombine(this.i.valueChanged, this._valueChanged_wrapped);
            ;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "actualMinimumValueChange", {
        get: function () {
            return this._actualMinimumValueChange;
        },
        set: function (ev) {
            var _this = this;
            if (this._actualMinimumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
                this._actualMinimumValueChange_wrapped = null;
                this._actualMinimumValueChange = null;
            }
            this._actualMinimumValueChange = ev;
            this._actualMinimumValueChange_wrapped = function (o, e) {
                var ext = _this.actualMinimumValue;
                if (e.propertyName == 'Bs') {
                    if (_this.beforeActualMinimumValueChange) {
                        _this.beforeActualMinimumValueChange(_this, ext);
                    }
                    if (_this._actualMinimumValueChange) {
                        _this._actualMinimumValueChange(_this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgcRadialGaugeComponent.prototype, "actualMaximumValueChange", {
        get: function () {
            return this._actualMaximumValueChange;
        },
        set: function (ev) {
            var _this = this;
            if (this._actualMaximumValueChange_wrapped !== null) {
                this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
                this._actualMaximumValueChange_wrapped = null;
                this._actualMaximumValueChange = null;
            }
            this._actualMaximumValueChange = ev;
            this._actualMaximumValueChange_wrapped = function (o, e) {
                var ext = _this.actualMaximumValue;
                if (e.propertyName == 'Br') {
                    if (_this.beforeActualMaximumValueChange) {
                        _this.beforeActualMaximumValueChange(_this, ext);
                    }
                    if (_this._actualMaximumValueChange) {
                        _this._actualMaximumValueChange(_this, ext);
                    }
                }
            };
            this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
        },
        enumerable: true,
        configurable: true
    });
    IgcRadialGaugeComponent._observedAttributesIgcRadialGaugeComponent = null;
    IgcRadialGaugeComponent.htmlTagName = "igc-radial-gauge";
    IgcRadialGaugeComponent._isElementRegistered = false;
    return IgcRadialGaugeComponent;
}(IgcHTMLElement));
export { IgcRadialGaugeComponent };
